{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tarea 2: Secuencias y formatos \n",
    "### <span style=\"color:blue\"> Eduardo Alejandro Arrieta Donato</span> \n",
    "## índice\n",
    "+ <a href=\"#Presentacion\" style=\"color:pink\">Presentación</a>\n",
    "+ <a href=\"#Ejercicio_1\" style=\"color:pink\">Ejercicio 1</a>\n",
    "+ <a href=\"#Ejercicio_2\" style=\"color:pink\">Ejercicio 2</a>\n",
    "+ <a href=\"#Ejercicio_3\" style=\"color:pink\">Ejercicio 3</a>\n",
    "+ <a href=\"#Ejercicio_4\" style=\"color:pink\">Ejercicio 4</a>\n",
    "+ <a href=\"#Ejercicio_5\" style=\"color:pink\">Ejercicio 5</a>\n",
    "+ <a href=\"#Ejercicio_6\" style=\"color:pink\">Ejercicio 6</a>\n",
    "+ <a href=\"#DONTCLICME\" style=\"color:pink\">DON´T CLICME</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Presentacion\"></a>\n",
    "## <span style=\"color:yellow\"> Presentación</span>\n",
    "Hola, esta es mi tarea2 del curso.\n",
    "Decidí enviarles la tarea como notebook ya que comentaron que iban a checar los ejercicios con sus propios archivos, así que no les complico la vida con otro formato.\n",
    "\n",
    "Cada ejercicio resulto viene con sus indicaciones generales y libertades que se tomaron.\n",
    "\n",
    "Cada celda ocupada para resolver tendrá sus respectiva documentación de código.\n",
    "\n",
    "Todos los $INPUTS$ serán *in-cell*, pues considero sucio hacerlo con `input()`, los inputos se destacaran con la leyenda <span style=\"color:orange\"> *'''INPUT'''*</span>\n",
    "\n",
    "Realmente no se como darle decoración a un notebook :c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_1\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_1</span>\n",
    "Un ORF inicia con un codón inicial y termina, ya sea con un codón final o al final de la cádena.\n",
    "\n",
    "Utilizando como input una cadena de ADN, obtener la cadena proteica de mayor longitud que puede ser traducida de cualquiera de sus ORF.\n",
    "\n",
    "Si más de una cadena protéica de longitud maxima existe, entonces podrás emitir cualquier solución."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">Para este ejerccio se pidió un como input una secuencia en str, no soy fan de hacer que se alargue la celda a los lados, pero es lo que pidió la actividad. Se decidió probar en todos los marcos de lectura y UNICAMENTE devolverá un restultado si se trata de un CDS. Para probar la celda usé el gen ascG de E. coli k-12 MG1655.</span>\n",
    "\n",
    "[ascG E. coli K-12 MG1655 NCBI](https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3?report=fasta&from=2838254&to=2839264&strand=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MTTMLEVAKRAGVSKATVSRVLSGNGYVSQETKDRVFQAVEESGYRPNLLARNLSAKSTQTLGLVVTNTLYHGIYFSELLFHAARMAEEKGRQLLLADGKHSAEEERQAIQYLLDLRCDAIMIYPRFLSVDEIDDIIDAHSQPIMVLNRRLRKNSSHSVWCDHKQTSFNAVAELINAGHQEIAFLTGSMDSPTSIERLAGYKDALAQHGIALNEKLIANGKWTPASGAEGVEMLLERGAKFSALVASNDDMAIGAMKALHERGVAVPEQVSVIGFDDIAIAPYTVPALSSVKIPVTEMIQEIIGRLIFMLDGGDFSPPKTFSGKLIRRDSLIAPSR*\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqUtils\n",
    "from Bio import SeqIO\n",
    "\n",
    "'''INPUT'''\n",
    "secuencia_input = 'ATGACGACGATGCTGGAAGTGGCGAAGCGCGCCGGGGTTTCAAAAGCGACCGTTTCCCGCGTGCTTTCAGGTAATGGCTACGTCAGCCAGGAGACTAAAGATCGCGTGTTTCAGGCGGTAGAAGAGAGCGGTTACCGTCCAAACTTGCTGGCGCGCAATCTGTCGGCGAAGAGTACTCAGACGCTGGGGCTGGTAGTGACCAACACGCTTTACCACGGCATTTATTTTAGTGAATTACTCTTTCATGCCGCGCGAATGGCGGAAGAGAAAGGGCGGCAGTTGCTATTGGCAGATGGTAAACACAGCGCAGAAGAAGAGCGCCAGGCGATTCAGTATCTGCTGGATCTGCGCTGCGACGCGATCATGATTTACCCGCGTTTTTTAAGCGTGGATGAGATCGATGACATCATTGACGCGCACAGTCAGCCGATAATGGTGCTTAATCGCCGCCTGCGCAAAAACAGCAGCCATAGCGTCTGGTGCGATCATAAACAGACCAGTTTTAACGCCGTGGCAGAGTTGATAAACGCCGGGCATCAGGAGATTGCTTTCCTTACCGGCTCGATGGATTCCCCCACCAGCATTGAACGTCTTGCCGGGTATAAAGACGCGCTGGCGCAGCATGGTATTGCGCTCAATGAAAAACTTATCGCTAACGGTAAATGGACGCCTGCCAGCGGGGCCGAAGGGGTAGAAATGTTGCTCGAACGTGGGGCTAAATTTAGCGCGTTAGTTGCCAGTAACGACGATATGGCGATAGGTGCGATGAAAGCGTTACACGAGCGCGGCGTAGCGGTGCCAGAGCAGGTGTCAGTTATCGGATTCGATGATATCGCTATTGCCCCCTACACCGTTCCGGCGCTCTCCAGCGTAAAAATTCCGGTGACTGAGATGATTCAGGAAATTATTGGACGGCTGATTTTTATGCTCGATGGCGGGGATTTCTCACCGCCGAAAACCTTCAGCGGAAAACTGATCCGCCGCGACTCACTCATTGCTCCTTCGCGATAA'\n",
    "\n",
    "# Genero una lista para todos los posibles péptidos en la secuencia forward\n",
    "peptidos_list = list()\n",
    "# Recorro cada marco de lectura\n",
    "for i in range(3):\n",
    "    # Aqui es donde se hace el dezplazamineto\n",
    "    secuencia = Seq(secuencia_input[i:])\n",
    "    # En caso de no encontrar una secuencia codificante...\n",
    "    try: \n",
    "        peptidos = secuencia.translate(cds = True)\n",
    "    except:\n",
    "        # ... Pasa al siguiente marco\n",
    "        continue\n",
    "    # Guardo todos los peptidos encontrados\n",
    "    peptidos = str(peptidos).split('*')\n",
    "    peptidos_list.extend(peptidos)\n",
    "\n",
    "# Recorro cada marco de lectura\n",
    "r_peptidos_list = list()    \n",
    "for i in range(3):\n",
    "    # Aqui es donde se hace el dezplazamineto en la secuencia reverse\n",
    "    secuencia = Seq(secuencia_input).reverse_complement()[i:]\n",
    "    # En caso de no encontrar una secuencia codificante...\n",
    "    try: \n",
    "        peptidos = secuencia.translate(cds = True)\n",
    "    except:\n",
    "        # ... Pasa al siguiente marco\n",
    "        continue\n",
    "    # Guardo todos los peptidos encontrados\n",
    "    peptidos = str(peptidos).split('*')\n",
    "    r_peptidos_list.extend(peptidos)\n",
    "    \n",
    "\n",
    "# Estos ifs son de control de errores, verifican que encontraron algun peptido\n",
    "if peptidos_list:\n",
    "    # Saca el de mayor longitud\n",
    "    proteina = max(peptidos_list)\n",
    "if r_peptidos_list:\n",
    "    # Saca el de mayor longitud en la forward\n",
    "    r_proteina = max(r_peptidos_list)\n",
    "\n",
    "    \n",
    "# Preparo mi secuencia para imprimirla\n",
    "mySequencia =  Seq(secuencia_input)\n",
    "# Control de errores\n",
    "if r_peptidos_list and peptidos_list and proteina < r_proteina:\n",
    "    # Si la proteina en reverse es más grande que la forward la imprime\n",
    "    mySequencia = mySequencia.reverse_complement()\n",
    "    print(f'Encontrada en reverse\\n {mySequencia.translate()}')\n",
    "    # Si no, imprime la forward\n",
    "else:\n",
    "    if peptidos_list:\n",
    "        print(mySequencia.translate())\n",
    "\n",
    "# Si no encuentra nada, imprime la siguiente cadena\n",
    "if not peptidos_list and not r_peptidos_list:\n",
    "    print('No se encontró CDS en ningún marco de lectura.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_2\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_2</span>\n",
    "    \n",
    "1. Teniendo como input un archivo fasta, crea un diccionario de todas las secuencias en tu atchivo FASTA.\n",
    "\n",
    "2. Encuentra todos los codones en tu primer marco de lectura. \n",
    "\n",
    "3. Imprime los codones de cada secuncia en formato FASTA. \n",
    "\n",
    "4. Imprime los codones de cada secuencia en los 6 marcos lectura."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">El ejercicio se dividió en dos celdas, la primera responde los incisos 1 a 3 y la segunda el 4. El output de la segunda es el completo porque así lo pide la actividad.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">seq1\n",
      "AAG AGC AGC TCG CGC TAA TGT GAT AGA TGG CGG TAA AGT AAA TGT \n",
      "CCT ATG GGC CAC CAA TTA TGG TGT ATG AGT GAA TCT CTG GTC CGA \n",
      "GAT TCA CTG AGT AAC TGC TGT ACA CAG TAG TAA CAC GTG GAG ATC \n",
      "CCA TAA GCT TCA CGT GTG GTC CAA TAA AAC ACT CCG TTG GTC AAC\n",
      "\n",
      ">seq2\n",
      "GCC ACA GAG CCT AGG ACC CCA ACC TAA CCT AAC CTA ACC TAA CCT \n",
      "ACA GTT TGA TCT TAA CCA TGA GGC TGA GAA GCG ATG TCC TGA CCG \n",
      "GCC TGT CCT AAC CGC CCT GAC CTA ACC GGC TTG ACC TAA CCG CCC \n",
      "TGA CCT AAC CAG GCT AAC CTA ACC AAA CCG TGA AAA AAG GAA TCT\n",
      "\n",
      ">seq3\n",
      "ATG AAA GTT ACA TAA AGA CTA TTC GAT GCA TAA ATA GTT CAG TTT \n",
      "TGA AAA CTT ACA TTT TGT TAA AGT CAG GTA CTT GTG TAT AAT ATC \n",
      "AAC TAA\n",
      "\n",
      ">seq4\n",
      "ATG CTA ACC AAA GTT TCA GTT CGG ACG TGT CGA TGA GCG ACG CTC \n",
      "AAA AAG GAA ACA ACA TGC CAA ATA GAA ACG ATC AAT TCG GCG ATG \n",
      "GAA ATC AGA ACA ACG ATC AGT TTG GAA ATC AAA ATA GAA ATA ACG \n",
      "GGA ACG ATC AGT TTA ATA ACA TGA TGC AGA ATA AAG GGA ATA ATC \n",
      "AAT TTA ATC CAG GTA ATC AGA ACA GAG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "'''INPUT'''\n",
    "path = 'seq.nt.fa'\n",
    "\n",
    "# Gnero un diccionario que almacene los records del fasta\n",
    "myDict = SeqIO.to_dict(SeqIO.parse(path, 'fasta'))\n",
    "\n",
    "# Recorro el diccionario\n",
    "for record in myDict.values():\n",
    "    # Extraigo cada codón en una lista\n",
    "    codones_list = re.findall('(.{3})', str(record.seq))\n",
    "    # Una lista y la convierto en objeto Seq\n",
    "    record.seq = Seq(' '.join(codones_list))\n",
    "    # Imprimo la entrada\n",
    "    print(record.format('fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Marco 1 forward\n",
      ">seq1\n",
      "AAG AGC AGC TCG CGC TAA TGT GAT AGA TGG CGG TAA AGT AAA TGT \n",
      "CCT ATG GGC CAC CAA TTA TGG TGT ATG AGT GAA TCT CTG GTC CGA \n",
      "GAT TCA CTG AGT AAC TGC TGT ACA CAG TAG TAA CAC GTG GAG ATC \n",
      "CCA TAA GCT TCA CGT GTG GTC CAA TAA AAC ACT CCG TTG GTC AAC\n",
      "\n",
      "Marco 1 forward\n",
      ">seq2\n",
      "GCC ACA GAG CCT AGG ACC CCA ACC TAA CCT AAC CTA ACC TAA CCT \n",
      "ACA GTT TGA TCT TAA CCA TGA GGC TGA GAA GCG ATG TCC TGA CCG \n",
      "GCC TGT CCT AAC CGC CCT GAC CTA ACC GGC TTG ACC TAA CCG CCC \n",
      "TGA CCT AAC CAG GCT AAC CTA ACC AAA CCG TGA AAA AAG GAA TCT\n",
      "\n",
      "Marco 1 forward\n",
      ">seq3\n",
      "ATG AAA GTT ACA TAA AGA CTA TTC GAT GCA TAA ATA GTT CAG TTT \n",
      "TGA AAA CTT ACA TTT TGT TAA AGT CAG GTA CTT GTG TAT AAT ATC \n",
      "AAC TAA\n",
      "\n",
      "Marco 1 forward\n",
      ">seq4\n",
      "ATG CTA ACC AAA GTT TCA GTT CGG ACG TGT CGA TGA GCG ACG CTC \n",
      "AAA AAG GAA ACA ACA TGC CAA ATA GAA ACG ATC AAT TCG GCG ATG \n",
      "GAA ATC AGA ACA ACG ATC AGT TTG GAA ATC AAA ATA GAA ATA ACG \n",
      "GGA ACG ATC AGT TTA ATA ACA TGA TGC AGA ATA AAG GGA ATA ATC \n",
      "AAT TTA ATC CAG GTA ATC AGA ACA GAG\n",
      "\n",
      "Marco 2 forward\n",
      ">seq1\n",
      "AGA GCA GCT CGC GCT AAT GTG ATA GAT GGC GGT AAA GTA AAT GTC \n",
      "CTA TGG GCC ACC AAT TAT GGT GTA TGA GTG AAT CTC TGG TCC GAG \n",
      "ATT CAC TGA GTA ACT GCT GTA CAC AGT AGT AAC ACG TGG AGA TCC \n",
      "CAT AAG CTT CAC GTG TGG TCC AAT AAA ACA CTC CGT TGG TCA\n",
      "\n",
      "Marco 2 forward\n",
      ">seq2\n",
      "CCA CAG AGC CTA GGA CCC CAA CCT AAC CTA ACC TAA CCT AAC CTA \n",
      "CAG TTT GAT CTT AAC CAT GAG GCT GAG AAG CGA TGT CCT GAC CGG \n",
      "CCT GTC CTA ACC GCC CTG ACC TAA CCG GCT TGA CCT AAC CGC CCT \n",
      "GAC CTA ACC AGG CTA ACC TAA CCA AAC CGT GAA AAA AGG AAT\n",
      "\n",
      "Marco 2 forward\n",
      ">seq3\n",
      "TGA AAG TTA CAT AAA GAC TAT TCG ATG CAT AAA TAG TTC AGT TTT \n",
      "GAA AAC TTA CAT TTT GTT AAA GTC AGG TAC TTG TGT ATA ATA TCA \n",
      "ACT AAA\n",
      "\n",
      "Marco 2 forward\n",
      ">seq4\n",
      "TGC TAA CCA AAG TTT CAG TTC GGA CGT GTC GAT GAG CGA CGC TCA \n",
      "AAA AGG AAA CAA CAT GCC AAA TAG AAA CGA TCA ATT CGG CGA TGG \n",
      "AAA TCA GAA CAA CGA TCA GTT TGG AAA TCA AAA TAG AAA TAA CGG \n",
      "GAA CGA TCA GTT TAA TAA CAT GAT GCA GAA TAA AGG GAA TAA TCA \n",
      "ATT TAA TCC AGG TAA TCA GAA CAG AGG\n",
      "\n",
      "Marco 3 forward\n",
      ">seq1\n",
      "GAG CAG CTC GCG CTA ATG TGA TAG ATG GCG GTA AAG TAA ATG TCC \n",
      "TAT GGG CCA CCA ATT ATG GTG TAT GAG TGA ATC TCT GGT CCG AGA \n",
      "TTC ACT GAG TAA CTG CTG TAC ACA GTA GTA ACA CGT GGA GAT CCC \n",
      "ATA AGC TTC ACG TGT GGT CCA ATA AAA CAC TCC GTT GGT CAA\n",
      "\n",
      "Marco 3 forward\n",
      ">seq2\n",
      "CAC AGA GCC TAG GAC CCC AAC CTA ACC TAA CCT AAC CTA ACC TAC \n",
      "AGT TTG ATC TTA ACC ATG AGG CTG AGA AGC GAT GTC CTG ACC GGC \n",
      "CTG TCC TAA CCG CCC TGA CCT AAC CGG CTT GAC CTA ACC GCC CTG \n",
      "ACC TAA CCA GGC TAA CCT AAC CAA ACC GTG AAA AAA GGA ATC\n",
      "\n",
      "Marco 3 forward\n",
      ">seq3\n",
      "GAA AGT TAC ATA AAG ACT ATT CGA TGC ATA AAT AGT TCA GTT TTG \n",
      "AAA ACT TAC ATT TTG TTA AAG TCA GGT ACT TGT GTA TAA TAT CAA \n",
      "CTA AAT\n",
      "\n",
      "Marco 3 forward\n",
      ">seq4\n",
      "GCT AAC CAA AGT TTC AGT TCG GAC GTG TCG ATG AGC GAC GCT CAA \n",
      "AAA GGA AAC AAC ATG CCA AAT AGA AAC GAT CAA TTC GGC GAT GGA \n",
      "AAT CAG AAC AAC GAT CAG TTT GGA AAT CAA AAT AGA AAT AAC GGG \n",
      "AAC GAT CAG TTT AAT AAC ATG ATG CAG AAT AAA GGG AAT AAT CAA \n",
      "TTT AAT CCA GGT AAT CAG AAC AGA GGT\n",
      "\n",
      "Marco 1 reverse\n",
      ">seq1\n",
      "GTT GAC CAA CGG AGT GTT TTA TTG GAC CAC ACG TGA AGC TTA TGG \n",
      "GAT CTC CAC GTG TTA CTA CTG TGT ACA GCA GTT ACT CAG TGA ATC \n",
      "TCG GAC CAG AGA TTC ACT CAT ACA CCA TAA TTG GTG GCC CAT AGG \n",
      "ACA TTT ACT TTA CCG CCA TCT ATC ACA TTA GCG CGA GCT GCT CTT\n",
      "\n",
      "Marco 1 reverse\n",
      ">seq2\n",
      "AGA TTC CTT TTT TCA CGG TTT GGT TAG GTT AGC CTG GTT AGG TCA \n",
      "GGG CGG TTA GGT CAA GCC GGT TAG GTC AGG GCG GTT AGG ACA GGC \n",
      "CGG TCA GGA CAT CGC TTC TCA GCC TCA TGG TTA AGA TCA AAC TGT \n",
      "AGG TTA GGT TAG GTT AGG TTA GGT TGG GGT CCT AGG CTC TGT GGC\n",
      "\n",
      "Marco 1 reverse\n",
      ">seq3\n",
      "ATT TAG TTG ATA TTA TAC ACA AGT ACC TGA CTT TAA CAA AAT GTA \n",
      "AGT TTT CAA AAC TGA ACT ATT TAT GCA TCG AAT AGT CTT TAT GTA \n",
      "ACT TTC\n",
      "\n",
      "Marco 1 reverse\n",
      ">seq4\n",
      "ACC TCT GTT CTG ATT ACC TGG ATT AAA TTG ATT ATT CCC TTT ATT \n",
      "CTG CAT CAT GTT ATT AAA CTG ATC GTT CCC GTT ATT TCT ATT TTG \n",
      "ATT TCC AAA CTG ATC GTT GTT CTG ATT TCC ATC GCC GAA TTG ATC \n",
      "GTT TCT ATT TGG CAT GTT GTT TCC TTT TTG AGC GTC GCT CAT CGA \n",
      "CAC GTC CGA ACT GAA ACT TTG GTT AGC\n",
      "\n",
      "Marco 2 reverse\n",
      ">seq1\n",
      "TTG ACC AAC GGA GTG TTT TAT TGG ACC ACA CGT GAA GCT TAT GGG \n",
      "ATC TCC ACG TGT TAC TAC TGT GTA CAG CAG TTA CTC AGT GAA TCT \n",
      "CGG ACC AGA GAT TCA CTC ATA CAC CAT AAT TGG TGG CCC ATA GGA \n",
      "CAT TTA CTT TAC CGC CAT CTA TCA CAT TAG CGC GAG CTG CTC\n",
      "\n",
      "Marco 2 reverse\n",
      ">seq2\n",
      "GAT TCC TTT TTT CAC GGT TTG GTT AGG TTA GCC TGG TTA GGT CAG \n",
      "GGC GGT TAG GTC AAG CCG GTT AGG TCA GGG CGG TTA GGA CAG GCC \n",
      "GGT CAG GAC ATC GCT TCT CAG CCT CAT GGT TAA GAT CAA ACT GTA \n",
      "GGT TAG GTT AGG TTA GGT TAG GTT GGG GTC CTA GGC TCT GTG\n",
      "\n",
      "Marco 2 reverse\n",
      ">seq3\n",
      "TTT AGT TGA TAT TAT ACA CAA GTA CCT GAC TTT AAC AAA ATG TAA \n",
      "GTT TTC AAA ACT GAA CTA TTT ATG CAT CGA ATA GTC TTT ATG TAA \n",
      "CTT TCA\n",
      "\n",
      "Marco 2 reverse\n",
      ">seq4\n",
      "CCT CTG TTC TGA TTA CCT GGA TTA AAT TGA TTA TTC CCT TTA TTC \n",
      "TGC ATC ATG TTA TTA AAC TGA TCG TTC CCG TTA TTT CTA TTT TGA \n",
      "TTT CCA AAC TGA TCG TTG TTC TGA TTT CCA TCG CCG AAT TGA TCG \n",
      "TTT CTA TTT GGC ATG TTG TTT CCT TTT TGA GCG TCG CTC ATC GAC \n",
      "ACG TCC GAA CTG AAA CTT TGG TTA GCA\n",
      "\n",
      "Marco 3 reverse\n",
      ">seq1\n",
      "TGA CCA ACG GAG TGT TTT ATT GGA CCA CAC GTG AAG CTT ATG GGA \n",
      "TCT CCA CGT GTT ACT ACT GTG TAC AGC AGT TAC TCA GTG AAT CTC \n",
      "GGA CCA GAG ATT CAC TCA TAC ACC ATA ATT GGT GGC CCA TAG GAC \n",
      "ATT TAC TTT ACC GCC ATC TAT CAC ATT AGC GCG AGC TGC TCT\n",
      "\n",
      "Marco 3 reverse\n",
      ">seq2\n",
      "ATT CCT TTT TTC ACG GTT TGG TTA GGT TAG CCT GGT TAG GTC AGG \n",
      "GCG GTT AGG TCA AGC CGG TTA GGT CAG GGC GGT TAG GAC AGG CCG \n",
      "GTC AGG ACA TCG CTT CTC AGC CTC ATG GTT AAG ATC AAA CTG TAG \n",
      "GTT AGG TTA GGT TAG GTT AGG TTG GGG TCC TAG GCT CTG TGG\n",
      "\n",
      "Marco 3 reverse\n",
      ">seq3\n",
      "TTA GTT GAT ATT ATA CAC AAG TAC CTG ACT TTA ACA AAA TGT AAG \n",
      "TTT TCA AAA CTG AAC TAT TTA TGC ATC GAA TAG TCT TTA TGT AAC \n",
      "TTT CAT\n",
      "\n",
      "Marco 3 reverse\n",
      ">seq4\n",
      "CTC TGT TCT GAT TAC CTG GAT TAA ATT GAT TAT TCC CTT TAT TCT \n",
      "GCA TCA TGT TAT TAA ACT GAT CGT TCC CGT TAT TTC TAT TTT GAT \n",
      "TTC CAA ACT GAT CGT TGT TCT GAT TTC CAT CGC CGA ATT GAT CGT \n",
      "TTC TAT TTG GCA TGT TGT TTC CTT TTT GAG CGT CGC TCA TCG ACA \n",
      "CGT CCG AAC TGA AAC TTT GGT TAG CAT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Inicializo el directorio\n",
    "myDict = SeqIO.to_dict(SeqIO.parse(path, 'fasta'))\n",
    "\n",
    "# Recorro cada marco\n",
    "# Los tres de foreard\n",
    "for i in range(3):\n",
    "    for record in myDict.values():\n",
    "        # Obtengo codones con el dezplazamineto\n",
    "        codones_list = re.findall('(.{3})', str(record.seq)[i:])\n",
    "        # Genero el objeto Seq\n",
    "        record.seq = Seq(' '.join(codones_list))\n",
    "        # Imprimo el fasta con su marco\n",
    "        print(f'Marco {i + 1} forward')\n",
    "        print(record.format('fasta'))\n",
    "    # Re inicializo el directorio para el siguiente marco\n",
    "    myDict = SeqIO.to_dict(SeqIO.parse(path, 'fasta'))\n",
    "\n",
    "# Los tres de reverse\n",
    "for i in range(3):\n",
    "    for record in myDict.values():\n",
    "        # Obtengo codones con el dezplazamineto\n",
    "        codones_list = re.findall('(.{3})', str(record.seq.reverse_complement())[i:])\n",
    "        # Genero el objeto Seq\n",
    "        record.seq = Seq(' '.join(codones_list))\n",
    "        # Imprimo el fasta con su marco\n",
    "        print(f'Marco {i + 1} reverse')\n",
    "        print(record.format('fasta'))\n",
    "    # Re inicializo el directorio para el siguiente marco\n",
    "    myDict = SeqIO.to_dict(SeqIO.parse(path, 'fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p id=\"Ejercicio_3\"></p>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_3</span>\n",
    "\n",
    "<p>Obtén el <b>número</b> de lecturas cuyo promedio de calidad esté por debajo del umbral dado.</p>\n",
    "<p><span class=\"given-return\">Input:</span> Un umbral de calidad, Junto con un fastq.</p>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">Aqui solo importé mean de statistics.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from statistics import mean\n",
    "\n",
    "'''INPUT'''\n",
    "path = 'sample.fastq'\n",
    "umbral = 38\n",
    "\n",
    "# Inicializo un contador para los que estén debajo del umbral\n",
    "coontador = 0\n",
    "# Leo y recorro el fastq por cada record\n",
    "for record in SeqIO.parse(path, 'fastq'):\n",
    "    # Obtengo el promedio de la calidad de la lectura y lo comparo con el umbral\n",
    "    if mean(record.letter_annotations[\"phred_quality\"]) < umbral:\n",
    "        # Aplico el contador\n",
    "        coontador += 1\n",
    "\n",
    "# Imprimo el resultado\n",
    "print(coontador)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_4\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_4</span>\n",
    "El archivo hepatitis.gb contiene información sobre la región de ADN que codifica HBx, una proteína viral de la hepatitis B multifuncional involucrada en la modulación de varias vías al interactuar directa o indirectamente con factores del hospedero (degradación de proteínas, apoptosis, transcripción, transducción de señales, progreso del ciclo celular y estabilidad genética). Escribe un programa de Python que, utilizando este registro de genbank, guarde la secuencia de proteínas correspondiente en formato fasta."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">Aproveché la función convert para generar un archivo fasta de un gb, y ya solo lo imprimo a pantalla.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">HE805982.1 Hepatitis B virus partial X gene for HBx, isolate 11851\n",
      "ATGGCTGCTAGGTTGTACTGCCAACTGGATTCTTCGCGGGACGTCCTTTGTTTACGTCCC\n",
      "GTCGGCGCTGAATCCCGCGGACGACCCCTCGCGAGGCCGCTTGGGGCTGTATCGTCCCCT\n",
      "TCTCCGTCTGCCGTACCGTCCGACCACGGGGCGCACCTCTCTTTACGCGGTCTCCCCGTC\n",
      "TGTGCCTTCTCATCTNGCCGGTCCGTGTGCACTTCGCTTCACCTCTGCACGTTGCATGGA\n",
      "GACCACCGTGAACGCCCATCARGTCCTGCCCAAGGTCTTACAAAAGAGGACTCTTGGACT\n",
      "CTCAGCAATGTCAACGACCGACCTAGAGGCCTACTTCAAAGACTGTGTGATAAAAGACTG\n",
      "GGAGGAGTTGGGGGAGGAGATAAGGTTAANNGTCTTTGTATTAGGAGGCTGTA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "'''INPUT'''\n",
    "path = 'hepatitis.gb'\n",
    "\n",
    "# Genero el path output para el archivo\n",
    "outputPath = re.sub('\\.gb', '.fa', path)\n",
    "\n",
    "# Genero el archivo output\n",
    "SeqIO.convert(path, 'genbank', outputPath, 'fasta')\n",
    "\n",
    "# Imprimo el record en formato fasta\n",
    "print(SeqIO.read(outputPath, 'fasta').format('fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_5\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_5</span>\n",
    "+ Consigue el *structure* del archivo pdb 4hhb\n",
    "\n",
    "+ Imprime las coordenadas 3d del átomo C-alpha del aminoacido con residuo número 20 en chain C y asigna ese residuo a una variable.\n",
    "\n",
    "+ Calcula el centro geométrico (centroid) del aminoacido mencionado anteriormente e imprímelo (esto asume que las masas de los átomos son iguales)\n",
    "\n",
    "Para calular el centro geométrico se necesitan sacar las coordenadas de todos los átomos del residuo y sacar el promedio."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">No hay mucho que decir, se describe con los comentarios.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La estructura: <Structure id=prot_4hhb>\n",
      "\n",
      "Cordenadas del CA del residuo HIS 20: [-26.912   8.767  11.705]\n",
      "\n",
      "Centro geométrico de HIS 20 = [-28.167499923706053, 8.430100059509277, 10.92150001525879]\n"
     ]
    }
   ],
   "source": [
    "from Bio import PDB\n",
    "\n",
    "'''INPUT'''\n",
    "path = '4hhb.pdb'\n",
    "\n",
    "# Genero el parseador\n",
    "parser = PDB.PDBParser(QUIET=True)\n",
    "\n",
    "# Obtengo la structura\n",
    "struc = parser.get_structure('prot_4hhb', path)\n",
    "\n",
    "# Imprimo el objeto struc\n",
    "print(f'La estructura: {struc}\\n')\n",
    "\n",
    "# Guardo el residuo y atomo objetivos\n",
    "residue = struc[0]['C'][20]\n",
    "calpha = residue['CA']\n",
    "\n",
    "# Obtengo sus coordenadas\n",
    "coord_capha = calpha.get_coord()\n",
    "# Impriimo estas características obtenidas\n",
    "print('Cordenadas del CA del residuo {} {}: {}\\n'.format(residue.get_resname(), \n",
    "                                                       residue.get_id()[1], \n",
    "                                                       coord_capha))\n",
    "\n",
    "# Para el centro geométrico inicializo tres int\n",
    "x = 0\n",
    "y = 0\n",
    "z = 0\n",
    "# Recorro cada atomo y sumo sus corrdenadas\n",
    "for atom in residue:\n",
    "    x += atom.get_coord()[0]\n",
    "    y += atom.get_coord()[1]\n",
    "    z += atom.get_coord()[2]\n",
    "\n",
    "# Obtengo el centro geometrico a partir de calcular su media\n",
    "natomos = len(residue)\n",
    "centroid = [x/natomos, y/natomos, z/natomos]\n",
    "\n",
    "# Imprimo los resultados\n",
    "print('Centro geométrico de {} {} = {}'.format(residue.get_resname(), \n",
    "                                               residue.get_id()[1], \n",
    "                                               centroid))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_6\"></a>\n",
    "## <span style=\"color:yellow\"> Ejercicio_6</span>\n",
    "\n",
    "1. Usando el archivo test.blastout.xml parsea los resultados e imprime lo siguiente \"queryName\thitID\te-value\thitDesc\tHsp_num\" Si el valor es mejor (meno) que 1e-10. (el query name lo encuentran en el blast_record y el hitDesc dentro del alignment.hit_def)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">Se parsea el docuemnto y se imprime lo que pide la actividad, restringí el output a solo tres alineamientos para no saturar en exceso.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query Name: Query_1\n",
      "Hit ID: sp|Q8CFZ5|S22AC_MOUSE\n",
      "e-Value: 1.47916e-109\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Mus musculus GN=Slc22a12 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q8CFZ5|S22AC_MOUSE\n",
      "e-Value: 1.20034e-74\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Mus musculus GN=Slc22a12 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q3ZAV1|S22AC_RAT\n",
      "e-Value: 7.69823e-103\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Rattus norvegicus GN=Slc22a12 PE=2 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q3ZAV1|S22AC_RAT\n",
      "e-Value: 1.31588e-62\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Rattus norvegicus GN=Slc22a12 PE=2 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q96S37|S22AC_HUMAN\n",
      "e-Value: 3.00718e-78\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Homo sapiens GN=SLC22A12 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q96S37|S22AC_HUMAN\n",
      "e-Value: 4.9569e-46\n",
      "Hit Description: Solute carrier family 22 member 12 OS=Homo sapiens GN=SLC22A12 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q9NSA0|S22AB_HUMAN\n",
      "e-Value: 4.379e-59\n",
      "Hit Description: Solute carrier family 22 member 11 OS=Homo sapiens GN=SLC22A11 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n",
      "Query Name: Query_1\n",
      "Hit ID: sp|Q9NSA0|S22AB_HUMAN\n",
      "e-Value: 6.63169e-25\n",
      "Hit Description: Solute carrier family 22 member 11 OS=Homo sapiens GN=SLC22A11 PE=1 SV=1\n",
      "Hsp Number: 2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "'''INPUT'''\n",
    "path = 'test.blastout.xml'\n",
    "\n",
    "texto = 'Query Name: {}\\nHit ID: {}\\ne-Value: {}\\nHit Description: {}\\nHsp Number: {}\\n'\n",
    "\n",
    "# Abro el archivo\n",
    "with open(path, 'r') as filexml:\n",
    "    # Lo parseo\n",
    "    blast_records = NCBIXML.parse(filexml)\n",
    "    # Recorro cada record\n",
    "    for blast_record in blast_records:\n",
    "        # Recorro cada alinieamineto\n",
    "        for i, alignment in enumerate(blast_record.alignments):\n",
    "            # Recorro cada característica hsp\n",
    "            for hsp in alignment.hsps:\n",
    "                \n",
    "                # Si el e-value no es bueno se ignora\n",
    "                if hsp.expect < 1e-10:\n",
    "                    # Imprimo las carasterísticas\n",
    "                    print(texto.format(\n",
    "                        \n",
    "                        blast_record.query_id,\n",
    "                        alignment.hit_id,\n",
    "                        hsp.expect,\n",
    "                        alignment.hit_def, \n",
    "                        len(alignment.hsps)\n",
    "                        \n",
    "                    ))\n",
    "            # Solo imprimo dee los tres primero alineamientos\n",
    "            if i > 2:\n",
    "                break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"DONTCLICLKME\"></a>\n",
    "## <span style=\"color:yellow\"> CHOMP!</span>\n",
    "![Te dije que le dieras clic](https://m.gifanimados.com/Gifs-Animales/Animaciones-Animales-Marinos/Peces/Tiburones/Boca-Tiburon-69162.gif)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
