{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tarea 4: ExPASy \n",
    "### <span style=\"color:blue\"> Eduardo Alejandro Arrieta Donato</span> \n",
    "## índice\n",
    "+ <a href=\"#Presentacion\" style=\"color:pink\">Presentación</a>\n",
    "+ <a href=\"#Ejercicio_1\" style=\"color:pink\">Ejercicio 1</a>\n",
    "+ <a href=\"#Ejercicio_2\" style=\"color:pink\">Ejercicio 2</a>\n",
    "+ <a href=\"#Ejercicio_3\" style=\"color:pink\">Ejercicio 3</a>\n",
    "+ <a href=\"#DONTCLICKME\" style=\"color:pink\">DON´T CLICKME</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Presentacion\"></a>\n",
    "## <span style=\"color:yellow\"> Presentación</span>\n",
    "Hola, esta es mi tarea4 del curso.\n",
    "\n",
    "Cada ejercicio resulto viene con sus indicaciones generales y libertades que se tomaron.\n",
    "\n",
    "Cada celda ocupada para resolver tendrá sus respectiva documentación de código."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <span style=\"color:red\"> README</span>\n",
    "Esta tarea me costó mucho trabajo, pues se me aglomeraron varias cosas, desde académicas como personales y de equipo. Pues me quedé sin mi computadora personal por un buen periodo de tiempo, apenas pude conseguirla de vuelta. No fue la única materia en la que tuve problemas.\n",
    "\n",
    "Muchas gracias por dejarme entregar la tarea después. Dos semanas de atraso si es una verdadera salvajada, por lo que acepto el castigo y penalización correspondiente."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_1\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_1</span>\n",
    "Busca los patrones de Prosite PS00079 y PS00080 en la Ferroxidasa humana HEPHL1 e imprme la o las partes de la secuencia que entra en el patron y el rango (donde inica y termina el patrón en la secuenca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['205785679', '148886707']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import Entrez \n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'earrieta@lcg.unam.mx'\n",
    "\n",
    "# Realizo una busqueda del gen en NCBI con el filtro para humano\n",
    "handle = Entrez.esearch(db=\"protein\", term='HEPHL1[Gene] AND Homo Sapiens[Orgn]') \n",
    "record_busquueda = Entrez.read(handle) \n",
    "handle.close()\n",
    "\n",
    "# Veo los resultados\n",
    "record_busquueda['IdList']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# Como son más de uno los itero y busco el que tenga referencia con UniProt\n",
    "for record in record_busquueda['IdList']:\n",
    "    # Obtengo su gb\n",
    "    handle_prot = Entrez.efetch(db=\"protein\", id=record, rettype=\"gb\", retmode=\"text\")\n",
    "    record_prot = SeqIO.read(handle_prot, \"genbank\")\n",
    "    handle_prot.close()\n",
    "    \n",
    "    if 'UniProtKB' in record_prot.annotations['db_source']:\n",
    "        # Extraigo su Entry de UniProt\n",
    "        ID_UniProt = re.findall('accession (.*);', record_prot.annotations['db_source'])\n",
    "        # Guardo el gb bueno\n",
    "        gb_proteina = record_prot\n",
    "    \n",
    "    # Pueden existir más de un link a Uniprot, pero por lo general no hay redundancia en SwissProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 404,
     "status": "ok",
     "timestamp": 1607840283835,
     "user": {
      "displayName": "Eduardo Arrieta",
      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14GgzqdjkhKSZ204Nx5Ykh7054SPeKxPMOPNi7jy5=s64",
      "userId": "01023628568219104715"
     },
     "user_tz": 360
    },
    "id": "TLuko4HZu6jF",
    "outputId": "ed66e12d-29ec-4032-e6b4-f63ee745d3db"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Q6MZM0']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ID_UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prosite\tRango\tSequencia\n",
      "PS00079\t341-361\tKWMITCQVSDHLQAGMLGQY\n",
      "PS00079\t694-714\tIFRVFCATMPHLSRGMGQIY\n",
      "PS00079\t1043-1063\tTWLLHCHVSDHIHAGMETTY\n",
      "PS00080\t1048-1059\tCHVSDHIHAGM\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import ScanProsite\n",
    "\n",
    "# Obtengo sus Prosites\n",
    "handle_PS = ScanProsite.scan(seq= ID_UniProt[0])\n",
    "result_PS = ScanProsite.read(handle_PS)\n",
    "\n",
    "# Estos son los objetivos a imprimir\n",
    "objetivos = ['PS00079','PS00080']\n",
    "\n",
    "# Header\n",
    "print('Prosite\\tRango\\tSequencia')\n",
    "for prosite in result_PS:\n",
    "    # Veo si están en mis objetivos\n",
    "    if prosite['signature_ac'] in objetivos:\n",
    "        # Obtengo su localización\n",
    "        start = prosite['start']\n",
    "        end = prosite['stop']\n",
    "        # Imprimo la información\n",
    "        print('{}\\t{}-{}'.format(prosite['signature_ac'], start, end), end= '\\t')\n",
    "        print(gb_proteina.seq[start:end])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_2\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_2</span>\n",
    "Obtén el nombre de las enzimas de una lista que contenga accesion numbers de SwissProt (los que quieras) utilizando expresiones regulares"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">La siguiente celda debe estra oculta, fue el parche de corrección para el modulo SwissProt de la librería de Biopython.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Copyright 2007 by Michiel de Hoon.  All rights reserved.\n",
    "# This code is part of the Biopython distribution and governed by its\n",
    "# license.  Please see the LICENSE file that should have been included\n",
    "# as part of this package.\n",
    "\"\"\"Code to work with the sprotXX.dat file from SwissProt.\n",
    "\n",
    "https://web.expasy.org/docs/userman.html\n",
    "\n",
    "Classes:\n",
    " - Record             Holds SwissProt data.\n",
    " - Reference          Holds reference data from a SwissProt record.\n",
    "\n",
    "Functions:\n",
    " - read               Read one SwissProt record\n",
    " - parse              Read multiple SwissProt records\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import io\n",
    "\n",
    "from Bio.SeqFeature import (\n",
    "    SeqFeature,\n",
    "    FeatureLocation,\n",
    "    ExactPosition,\n",
    "    BeforePosition,\n",
    "    AfterPosition,\n",
    "    UncertainPosition,\n",
    "    UnknownPosition,\n",
    ")\n",
    "\n",
    "\n",
    "class SwissProtParserError(ValueError):\n",
    "    \"\"\"An error occurred while parsing a SwissProt file.\"\"\"\n",
    "\n",
    "    def __init__(self, *args, line=None):\n",
    "        \"\"\"Create a SwissProtParserError object with the offending line.\"\"\"\n",
    "        super().__init__(*args)\n",
    "        self.line = line\n",
    "\n",
    "\n",
    "class Record:\n",
    "    \"\"\"Holds information from a SwissProt record.\n",
    "\n",
    "    Attributes:\n",
    "     - entry_name        Name of this entry, e.g. RL1_ECOLI.\n",
    "     - data_class        Either 'STANDARD' or 'PRELIMINARY'.\n",
    "     - molecule_type     Type of molecule, 'PRT',\n",
    "     - sequence_length   Number of residues.\n",
    "     - accessions        List of the accession numbers, e.g. ['P00321']\n",
    "     - created           A tuple of (date, release).\n",
    "     - sequence_update   A tuple of (date, release).\n",
    "     - annotation_update A tuple of (date, release).\n",
    "     - description       Free-format description.\n",
    "     - gene_name         Gene name.  See userman.txt for description.\n",
    "     - organism          The source of the sequence.\n",
    "     - organelle         The origin of the sequence.\n",
    "     - organism_classification  The taxonomy classification.  List of strings.\n",
    "       (http://www.ncbi.nlm.nih.gov/Taxonomy/)\n",
    "     - taxonomy_id       A list of NCBI taxonomy id's.\n",
    "     - host_organism     A list of names of the hosts of a virus, if any.\n",
    "     - host_taxonomy_id  A list of NCBI taxonomy id's of the hosts, if any.\n",
    "     - references        List of Reference objects.\n",
    "     - comments          List of strings.\n",
    "     - cross_references  List of tuples (db, id1[, id2][, id3]).  See the docs.\n",
    "     - keywords          List of the keywords.\n",
    "     - features          List of tuples (key name, from, to, description).\n",
    "       from and to can be either integers for the residue\n",
    "       numbers, '<', '>', or '?'\n",
    "     - protein_existence Numerical value describing the evidence for the existence of the protein.\n",
    "     - seqinfo           tuple of (length, molecular weight, CRC32 value)\n",
    "     - sequence          The sequence.\n",
    "\n",
    "    Examples\n",
    "    --------\n",
    "    >>> from Bio import SwissProt\n",
    "    >>> example_filename = \"SwissProt/sp008\"\n",
    "    >>> with open(example_filename) as handle:\n",
    "    ...     records = SwissProt.parse(handle)\n",
    "    ...     for record in records:\n",
    "    ...         print(record.entry_name)\n",
    "    ...         print(\",\".join(record.accessions))\n",
    "    ...         print(record.keywords)\n",
    "    ...         print(repr(record.organism))\n",
    "    ...         print(record.sequence[:20] + \"...\")\n",
    "    ...\n",
    "    1A02_HUMAN\n",
    "    P01892,P06338,P30514,P30444,P30445,P30446,Q29680,Q29899,Q95352,Q29837,Q95380\n",
    "    ['MHC I', 'Transmembrane', 'Glycoprotein', 'Signal', 'Polymorphism', '3D-structure']\n",
    "    'Homo sapiens (Human).'\n",
    "    MAVMAPRTLVLLLSGALALT...\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        \"\"\"Initialize the class.\"\"\"\n",
    "        self.entry_name = None\n",
    "        self.data_class = None\n",
    "        self.molecule_type = None\n",
    "        self.sequence_length = None\n",
    "\n",
    "        self.accessions = []\n",
    "        self.created = None\n",
    "        self.sequence_update = None\n",
    "        self.annotation_update = None\n",
    "\n",
    "        self.description = []\n",
    "        self.gene_name = \"\"\n",
    "        self.organism = []\n",
    "        self.organelle = \"\"\n",
    "        self.organism_classification = []\n",
    "        self.taxonomy_id = []\n",
    "        self.host_organism = []\n",
    "        self.host_taxonomy_id = []\n",
    "        self.references = []\n",
    "        self.comments = []\n",
    "        self.cross_references = []\n",
    "        self.keywords = []\n",
    "        self.features = []\n",
    "        self.protein_existence = \"\"\n",
    "\n",
    "        self.seqinfo = None\n",
    "        self.sequence = \"\"\n",
    "\n",
    "\n",
    "class Reference:\n",
    "    \"\"\"Holds information from one reference in a SwissProt entry.\n",
    "\n",
    "    Attributes:\n",
    "     - number      Number of reference in an entry.\n",
    "     - evidence    Evidence code.  List of strings.\n",
    "     - positions   Describes extent of work.  List of strings.\n",
    "     - comments    Comments.  List of (token, text).\n",
    "     - references  References.  List of (dbname, identifier).\n",
    "     - authors     The authors of the work.\n",
    "     - title       Title of the work.\n",
    "     - location    A citation for the work.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        \"\"\"Initialize the class.\"\"\"\n",
    "        self.number = None\n",
    "        self.positions = []\n",
    "        self.comments = []\n",
    "        self.references = []\n",
    "        self.authors = []\n",
    "        self.title = []\n",
    "        self.location = []\n",
    "\n",
    "\n",
    "class FeatureTable(SeqFeature):\n",
    "    \"\"\"Stores feature annotations for specific regions of the sequence.\n",
    "\n",
    "    This is a subclass of SeqFeature, defined in Bio.SeqFeature, where the\n",
    "    attributes are used as follows:\n",
    "\n",
    "     - ``location``: location of the feature on the canonical or isoform\n",
    "       sequence; the location is stored as an instance of FeatureLocation,\n",
    "       defined in Bio.SeqFeature, with the ref attribute set to the isoform\n",
    "       ID referring to the canonical or isoform sequence on which the feature\n",
    "       is defined\n",
    "     - ``id``: unique and stable identifier (FTId), only provided for features\n",
    "       belonging to the types CARBOHYD, CHAIN, PEPTIDE, PROPEP, VARIANT, or\n",
    "       VAR_SEQ\n",
    "     - ``type``: indicates the type of feature, as defined by the UniProt\n",
    "       Knowledgebase documentation:\n",
    "\n",
    "        - ACT_SITE: amino acid(s) involved in the activity of an enzyme\n",
    "        - BINDING:  binding site for any chemical group\n",
    "        - CARBOHYD: glycosylation site; an FTId identifier to the GlyConnect\n",
    "          database is provided if annotated there\n",
    "        - CA_BIND:  calcium-binding region\n",
    "        - CHAIN:    polypeptide chain in the mature protein\n",
    "        - COILED:   coiled-coil region\n",
    "        - COMPBIAS: compositionally biased region\n",
    "        - CONFLICT: different sources report differing sequences\n",
    "        - CROSSLNK: posttransationally formed amino acid bond\n",
    "        - DISULFID: disulfide bond\n",
    "        - DNA_BIND: DNA-binding region\n",
    "        - DOMAIN:   domain, defined as a specific combination of secondary\n",
    "          structures organized into a characteristic three-dimensional\n",
    "          structure or fold\n",
    "        - INIT_MET: initiator methionine\n",
    "        - INTRAMEM: region located in a membrane without crossing it\n",
    "        - HELIX:    alpha-, 3(10)-, or pi-helix secondary structure\n",
    "        - LIPID:    covalent binding of a lipid moiety\n",
    "        - METAL:    binding site for a metal ion\n",
    "        - MOD_RES:  posttranslational modification (PTM) of a residue,\n",
    "          annotated by the controlled vocabulary defined by the ptmlist.txt\n",
    "          document on the UniProt website\n",
    "        - MOTIF:    short sequence motif of biological interest\n",
    "        - MUTAGEN:  site experimentally altered by mutagenesis\n",
    "        - NON_CONS: non-consecutive residues\n",
    "        - NON_STD:  non-standard amino acid\n",
    "        - NON_TER:  the residue at an extremity of the sequence is not the\n",
    "          terminal residue\n",
    "        - NP_BIND:  nucleotide phosphate-binding region\n",
    "        - PEPTIDE:  released active mature polypeptide\n",
    "        - PROPEP:   any processed propeptide\n",
    "        - REGION:   region of interest in the sequence\n",
    "        - REPEAT:   internal sequence repetition\n",
    "        - SIGNAL:   signal sequence (prepeptide)\n",
    "        - SITE:     amino-acid site of interest not represented by another\n",
    "          feature key\n",
    "        - STRAND:   beta-strand secondary structure; either a hydrogen-bonded\n",
    "          extended beta strand or a residue in an isolated beta-bridge\n",
    "        - TOPO_DOM: topological domain\n",
    "        - TRANSIT:  transit peptide (mitochondrion, chloroplast, thylakoid,\n",
    "          cyanelle, peroxisome, etc.)\n",
    "        - TRANSMEM: transmembrane region\n",
    "        - TURN:     H-bonded turn (3-, 4-, or 5-turn)\n",
    "        - UNSURE:   uncertainties in the sequence\n",
    "        - VARIANT:  sequence variant; an FTId is provided for protein sequence\n",
    "          variants of Hominidae (great apes and humans)\n",
    "        - VAR_SEQ:  sequence variant produced by alternative splicing,\n",
    "          alternative promoter usage, alternative initiation, or ribosomal\n",
    "          frameshifting\n",
    "        - ZN_FING:  zinc finger region\n",
    "\n",
    "     - qualifiers   A dictionary of additional information, which may include\n",
    "       the feature evidence and free-text notes. While SwissProt includes the\n",
    "       feature identifier code (FTId) as a qualifier, it is stored as the\n",
    "       attribute ID of the FeatureTable object.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "def parse(source):\n",
    "    \"\"\"Read multiple SwissProt records from file.\n",
    "\n",
    "    Argument source is a file-like object or a path to a file.\n",
    "\n",
    "    Returns a generator object which yields Bio.SwissProt.Record() objects.\n",
    "    \"\"\"\n",
    "    handle = _open(source)\n",
    "    try:\n",
    "        while True:\n",
    "            record = _read(handle)\n",
    "            if not record:\n",
    "                return\n",
    "            yield record\n",
    "    finally:\n",
    "        if handle is not source:\n",
    "            handle.close()\n",
    "\n",
    "\n",
    "def read(source):\n",
    "    \"\"\"Read one SwissProt record from file.\n",
    "\n",
    "    Argument source is a file-like object or a path to a file.\n",
    "\n",
    "    Returns a Record() object.\n",
    "    \"\"\"\n",
    "    handle = _open(source)\n",
    "    try:\n",
    "        record = _read(handle)\n",
    "        if not record:\n",
    "            raise ValueError(\"No SwissProt record found\")\n",
    "        # We should have reached the end of the record by now.\n",
    "        # Try to read one more line to be sure:\n",
    "        try:\n",
    "            next(handle)\n",
    "        except StopIteration:\n",
    "            return record\n",
    "        raise ValueError(\"More than one SwissProt record found\")\n",
    "    finally:\n",
    "        if handle is not source:\n",
    "            handle.close()\n",
    "\n",
    "\n",
    "# Everything below is considered private\n",
    "\n",
    "\n",
    "def _open(source):\n",
    "    try:\n",
    "        handle = open(source)\n",
    "        return handle\n",
    "    except TypeError:\n",
    "        handle = source\n",
    "        if handle.read(0) == \"\":\n",
    "            # handle is text; assume the encoding is compatible with ASCII\n",
    "            return handle\n",
    "        # handle is binary; SwissProt encoding is always ASCII\n",
    "        return io.TextIOWrapper(handle, encoding=\"ASCII\")\n",
    "\n",
    "\n",
    "def _read(handle):\n",
    "    record = None\n",
    "    unread = \"\"\n",
    "    try:\n",
    "        line = next(handle)\n",
    "    except StopIteration:\n",
    "        return record\n",
    "    key, value = line[:2], line[5:].rstrip()\n",
    "    if key != \"ID\":\n",
    "        raise SwissProtParserError(\"Failed to find ID in first line\", line=line)\n",
    "    record = Record()\n",
    "    _read_id(record, line)\n",
    "    _sequence_lines = []\n",
    "    for line in handle:\n",
    "        key, value = line[:2], line[5:].rstrip()\n",
    "        if unread:\n",
    "            value = unread + \" \" + value\n",
    "            unread = \"\"\n",
    "        if key == \"AC\":\n",
    "            accessions = value.rstrip(\";\").split(\"; \")\n",
    "            record.accessions.extend(accessions)\n",
    "        elif key == \"DT\":\n",
    "            _read_dt(record, line)\n",
    "        elif key == \"DE\":\n",
    "            record.description.append(value.strip())\n",
    "        elif key == \"GN\":\n",
    "            if record.gene_name:\n",
    "                record.gene_name += \" \"\n",
    "            record.gene_name += value\n",
    "        elif key == \"OS\":\n",
    "            record.organism.append(value)\n",
    "        elif key == \"OG\":\n",
    "            record.organelle += line[5:]\n",
    "        elif key == \"OC\":\n",
    "            cols = value.rstrip(\";.\").split(\"; \")\n",
    "            record.organism_classification.extend(cols)\n",
    "        elif key == \"OX\":\n",
    "            _read_ox(record, line)\n",
    "        elif key == \"OH\":\n",
    "            _read_oh(record, line)\n",
    "        elif key == \"RN\":\n",
    "            reference = Reference()\n",
    "            _read_rn(reference, value)\n",
    "            record.references.append(reference)\n",
    "        elif key == \"RP\":\n",
    "            assert record.references, \"RP: missing RN\"\n",
    "            record.references[-1].positions.append(value)\n",
    "        elif key == \"RC\":\n",
    "            assert record.references, \"RC: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            unread = _read_rc(reference, value)\n",
    "        elif key == \"RX\":\n",
    "            assert record.references, \"RX: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            _read_rx(reference, value)\n",
    "        elif key == \"RL\":\n",
    "            assert record.references, \"RL: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.location.append(value)\n",
    "        # In UniProt release 1.12 of 6/21/04, there is a new RG\n",
    "        # (Reference Group) line, which references a group instead of\n",
    "        # an author.  Each block must have at least 1 RA or RG line.\n",
    "        elif key == \"RA\":\n",
    "            assert record.references, \"RA: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.authors.append(value)\n",
    "        elif key == \"RG\":\n",
    "            assert record.references, \"RG: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.authors.append(value)\n",
    "        elif key == \"RT\":\n",
    "            assert record.references, \"RT: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.title.append(value)\n",
    "        elif key == \"CC\":\n",
    "            _read_cc(record, line)\n",
    "        elif key == \"DR\":\n",
    "            _read_dr(record, value)\n",
    "        elif key == \"PE\":\n",
    "            _read_pe(record, value)\n",
    "        elif key == \"KW\":\n",
    "            _read_kw(record, value)\n",
    "        elif key == \"FT\":\n",
    "            _read_ft(record, line)\n",
    "        elif key == \"SQ\":\n",
    "            cols = value.split()\n",
    "            assert len(cols) == 7, \"I don't understand SQ line %s\" % line\n",
    "            # Do more checking here?\n",
    "            record.seqinfo = int(cols[1]), int(cols[3]), cols[5]\n",
    "        elif key == \"  \":\n",
    "            _sequence_lines.append(value.replace(\" \", \"\").rstrip())\n",
    "        elif key == \"//\":\n",
    "            # Join multiline data into one string\n",
    "            record.description = \" \".join(record.description)\n",
    "            record.organism = \" \".join(record.organism)\n",
    "            record.organelle = record.organelle.rstrip()\n",
    "            for reference in record.references:\n",
    "                reference.authors = \" \".join(reference.authors).rstrip(\";\")\n",
    "                if reference.title:\n",
    "                    title = reference.title[0]\n",
    "                    for fragment in reference.title[1:]:\n",
    "                        if not title.endswith(\"-\"):\n",
    "                            title += \" \"\n",
    "                        title += fragment\n",
    "                    title = title.rstrip(\";\")\n",
    "                    if title.startswith('\"') and title.endswith('\"'):\n",
    "                        title = title[1:-1]  # remove quotes\n",
    "                else:\n",
    "                    title = \"\"\n",
    "                reference.title = title\n",
    "                reference.location = \" \".join(reference.location)\n",
    "            record.sequence = \"\".join(_sequence_lines)\n",
    "            return record\n",
    "        elif key == \"**\":\n",
    "            # Do this one last, as it will almost never occur.\n",
    "            # See Bug 2353, some files from the EBI have extra lines\n",
    "            # starting \"**\" (two asterisks/stars).  They appear\n",
    "            # to be unofficial automated annotations. e.g.\n",
    "            # **\n",
    "            # **   #################    INTERNAL SECTION    ##################\n",
    "            # **HA SAM; Annotated by PicoHamap 1.88; MF_01138.1; 09-NOV-2003.\n",
    "            pass\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unknown keyword '%s' found\" % key, line=line)\n",
    "    if record:\n",
    "        raise ValueError(\"Unexpected end of stream.\")\n",
    "\n",
    "\n",
    "def _read_id(record, line):\n",
    "    cols = line[5:].split()\n",
    "    # Prior to release 51, included with MoleculeType:\n",
    "    # ID   EntryName DataClass; MoleculeType; SequenceLength AA.\n",
    "    #\n",
    "    # Newer files lack the MoleculeType:\n",
    "    # ID   EntryName DataClass; SequenceLength AA.\n",
    "    if len(cols) == 5:\n",
    "        record.entry_name = cols[0]\n",
    "        record.data_class = cols[1].rstrip(\";\")\n",
    "        record.molecule_type = cols[2].rstrip(\";\")\n",
    "        record.sequence_length = int(cols[3])\n",
    "    elif len(cols) == 4:\n",
    "        record.entry_name = cols[0]\n",
    "        record.data_class = cols[1].rstrip(\";\")\n",
    "        record.molecule_type = None\n",
    "        record.sequence_length = int(cols[2])\n",
    "    else:\n",
    "        raise SwissProtParserError(\"ID line has unrecognised format\", line=line)\n",
    "    # check if the data class is one of the allowed values\n",
    "    allowed = (\"STANDARD\", \"PRELIMINARY\", \"IPI\", \"Reviewed\", \"Unreviewed\")\n",
    "    if record.data_class not in allowed:\n",
    "        message = \"Unrecognized data class '%s'\" % record.data_class\n",
    "        raise SwissProtParserError(message, line=line)\n",
    "\n",
    "    # molecule_type should be 'PRT' for PRoTein\n",
    "    # Note that has been removed in recent releases (set to None)\n",
    "    if record.molecule_type not in (None, \"PRT\"):\n",
    "        message = \"Unrecognized molecule type '%s'\" % record.molecule_type\n",
    "        raise SwissProtParserError(message, line=line)\n",
    "\n",
    "\n",
    "def _read_dt(record, line):\n",
    "    value = line[5:]\n",
    "    uprline = value.upper()\n",
    "    cols = value.rstrip().split()\n",
    "    if (\n",
    "        \"CREATED\" in uprline\n",
    "        or \"LAST SEQUENCE UPDATE\" in uprline\n",
    "        or \"LAST ANNOTATION UPDATE\" in uprline\n",
    "    ):\n",
    "        # Old style DT line\n",
    "        # =================\n",
    "        # e.g.\n",
    "        # DT   01-FEB-1995 (Rel. 31, Created)\n",
    "        # DT   01-FEB-1995 (Rel. 31, Last sequence update)\n",
    "        # DT   01-OCT-2000 (Rel. 40, Last annotation update)\n",
    "        #\n",
    "        # or:\n",
    "        # DT   08-JAN-2002 (IPI Human rel. 2.3, Created)\n",
    "        # ...\n",
    "\n",
    "        # find where the version information will be located\n",
    "        # This is needed for when you have cases like IPI where\n",
    "        # the release version is in a different spot:\n",
    "        # DT   08-JAN-2002 (IPI Human rel. 2.3, Created)\n",
    "        uprcols = uprline.split()\n",
    "        rel_index = -1\n",
    "        for index in range(len(uprcols)):\n",
    "            if \"REL.\" in uprcols[index]:\n",
    "                rel_index = index\n",
    "        assert rel_index >= 0, \"Could not find Rel. in DT line: %s\" % line\n",
    "        version_index = rel_index + 1\n",
    "        # get the version information\n",
    "        str_version = cols[version_index].rstrip(\",\")\n",
    "        # no version number\n",
    "        if str_version == \"\":\n",
    "            version = 0\n",
    "        # dot versioned\n",
    "        elif \".\" in str_version:\n",
    "            version = str_version\n",
    "        # integer versioned\n",
    "        else:\n",
    "            version = int(str_version)\n",
    "        date = cols[0]\n",
    "\n",
    "        if \"CREATED\" in uprline:\n",
    "            record.created = date, version\n",
    "        elif \"LAST SEQUENCE UPDATE\" in uprline:\n",
    "            record.sequence_update = date, version\n",
    "        elif \"LAST ANNOTATION UPDATE\" in uprline:\n",
    "            record.annotation_update = date, version\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unrecognised DT (DaTe) line\", line=line)\n",
    "    elif (\n",
    "        \"INTEGRATED INTO\" in uprline\n",
    "        or \"SEQUENCE VERSION\" in uprline\n",
    "        or \"ENTRY VERSION\" in uprline\n",
    "    ):\n",
    "        # New style DT line\n",
    "        # =================\n",
    "        # As of UniProt Knowledgebase release 7.0 (including\n",
    "        # Swiss-Prot release 49.0 and TrEMBL release 32.0) the\n",
    "        # format of the DT lines and the version information\n",
    "        # in them was changed - the release number was dropped.\n",
    "        #\n",
    "        # For more information see bug 1948 and\n",
    "        # http://ca.expasy.org/sprot/relnotes/sp_news.html#rel7.0\n",
    "        #\n",
    "        # e.g.\n",
    "        # DT   01-JAN-1998, integrated into UniProtKB/Swiss-Prot.\n",
    "        # DT   15-OCT-2001, sequence version 3.\n",
    "        # DT   01-APR-2004, entry version 14.\n",
    "        #\n",
    "        # This is a new style DT line...\n",
    "\n",
    "        # The date should be in string cols[1]\n",
    "        # Get the version number if there is one.\n",
    "        # For the three DT lines above: 0, 3, 14\n",
    "        try:\n",
    "            version = 0\n",
    "            for s in cols[-1].split(\".\"):\n",
    "                if s.isdigit():\n",
    "                    version = int(s)\n",
    "        except ValueError:\n",
    "            version = 0\n",
    "        date = cols[0].rstrip(\",\")\n",
    "\n",
    "        # Re-use the historical property names, even though\n",
    "        # the meaning has changed slighty:\n",
    "        if \"INTEGRATED\" in uprline:\n",
    "            record.created = date, version\n",
    "        elif \"SEQUENCE VERSION\" in uprline:\n",
    "            record.sequence_update = date, version\n",
    "        elif \"ENTRY VERSION\" in uprline:\n",
    "            record.annotation_update = date, version\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unrecognised DT (DaTe) line\", line=line)\n",
    "    else:\n",
    "        raise SwissProtParserError(\"Failed to parse DT (DaTe) line\", line=line)\n",
    "\n",
    "\n",
    "def _read_ox(record, line):\n",
    "    # The OX line used to be in the simple format:\n",
    "    # OX   DESCRIPTION=ID[, ID]...;\n",
    "    # If there are too many id's to fit onto a line, then the ID's\n",
    "    # continue directly onto the next line, e.g.\n",
    "    # OX   DESCRIPTION=ID[, ID]...\n",
    "    # OX   ID[, ID]...;\n",
    "    # Currently, the description is always \"NCBI_TaxID\".\n",
    "    # To parse this, I need to check to see whether I'm at the\n",
    "    # first line.  If I am, grab the description and make sure\n",
    "    # it's an NCBI ID.  Then, grab all the id's.\n",
    "    #\n",
    "    # As of the 2014-10-01 release, there may be an evidence code, e.g.\n",
    "    # OX   NCBI_TaxID=418404 {ECO:0000313|EMBL:AEX14553.1};\n",
    "    # In the short term, we will ignore any evidence codes:\n",
    "    line = line.split(\"{\")[0]\n",
    "    if record.taxonomy_id:\n",
    "        ids = line[5:].rstrip().rstrip(\";\")\n",
    "    else:\n",
    "        descr, ids = line[5:].rstrip().rstrip(\";\").split(\"=\")\n",
    "        assert descr == \"NCBI_TaxID\", \"Unexpected taxonomy type %s\" % descr\n",
    "    record.taxonomy_id.extend(ids.split(\", \"))\n",
    "\n",
    "\n",
    "def _read_oh(record, line):\n",
    "    # Line type OH (Organism Host) for viral hosts\n",
    "    assert line[5:].startswith(\"NCBI_TaxID=\"), \"Unexpected %s\" % line\n",
    "    line = line[16:].rstrip()\n",
    "    assert line[-1] == \".\" and line.count(\";\") == 1, line\n",
    "    taxid, name = line[:-1].split(\";\")\n",
    "    record.host_taxonomy_id.append(taxid.strip())\n",
    "    record.host_organism.append(name.strip())\n",
    "\n",
    "\n",
    "def _read_rn(reference, rn):\n",
    "    # This used to be a very simple line with a reference number, e.g.\n",
    "    # RN   [1]\n",
    "    # As of the 2014-10-01 release, there may be an evidence code, e.g.\n",
    "    # RN   [1] {ECO:0000313|EMBL:AEX14553.1}\n",
    "    words = rn.split(None, 1)\n",
    "    number = words[0]\n",
    "    assert number.startswith(\"[\") and number.endswith(\"]\"), (\n",
    "        \"Missing brackets %s\" % number\n",
    "    )\n",
    "    reference.number = int(number[1:-1])\n",
    "    if len(words) > 1:\n",
    "        evidence = words[1]\n",
    "        assert evidence.startswith(\"{\") and evidence.endswith(\"}\"), (\n",
    "            \"Missing braces %s\" % evidence\n",
    "        )\n",
    "        reference.evidence = evidence[1:-1].split(\"|\")\n",
    "\n",
    "\n",
    "def _read_rc(reference, value):\n",
    "    cols = value.split(\";\")\n",
    "    if value[-1] == \";\":\n",
    "        unread = \"\"\n",
    "    else:\n",
    "        cols, unread = cols[:-1], cols[-1]\n",
    "    for col in cols:\n",
    "        if not col:  # last column will be the empty string\n",
    "            return\n",
    "        # The token is everything before the first '=' character.\n",
    "        i = col.find(\"=\")\n",
    "        if i >= 0:\n",
    "            token, text = col[:i], col[i + 1 :]\n",
    "            comment = token.lstrip(), text\n",
    "            reference.comments.append(comment)\n",
    "        else:\n",
    "            comment = reference.comments[-1]\n",
    "            comment = \"%s %s\" % (comment, col)\n",
    "            reference.comments[-1] = comment\n",
    "    return unread\n",
    "\n",
    "\n",
    "def _read_rx(reference, value):\n",
    "    # The basic (older?) RX line is of the form:\n",
    "    # RX   MEDLINE; 85132727.\n",
    "    # but there are variants of this that need to be dealt with (see below)\n",
    "\n",
    "    # CLD1_HUMAN in Release 39 and DADR_DIDMA in Release 33\n",
    "    # have extraneous information in the RX line.  Check for\n",
    "    # this and chop it out of the line.\n",
    "    # (noticed by katel@worldpath.net)\n",
    "    value = value.replace(\" [NCBI, ExPASy, Israel, Japan]\", \"\")\n",
    "\n",
    "    # RX lines can also be used of the form\n",
    "    # RX   PubMed=9603189;\n",
    "    # reported by edvard@farmasi.uit.no\n",
    "    # and these can be more complicated like:\n",
    "    # RX   MEDLINE=95385798; PubMed=7656980;\n",
    "    # RX   PubMed=15060122; DOI=10.1136/jmg 2003.012781;\n",
    "    # We look for these cases first and deal with them\n",
    "    warn = False\n",
    "    if \"=\" in value:\n",
    "        cols = value.split(\"; \")\n",
    "        cols = [x.strip() for x in cols]\n",
    "        cols = [x for x in cols if x]\n",
    "        for col in cols:\n",
    "            x = col.split(\"=\")\n",
    "            if len(x) != 2 or x == (\"DOI\", \"DOI\"):\n",
    "                warn = True\n",
    "                break\n",
    "            assert len(x) == 2, \"I don't understand RX line %s\" % value\n",
    "            reference.references.append((x[0], x[1].rstrip(\";\")))\n",
    "    # otherwise we assume we have the type 'RX   MEDLINE; 85132727.'\n",
    "    else:\n",
    "        cols = value.split(\"; \")\n",
    "        # normally we split into the three parts\n",
    "        if len(cols) != 2:\n",
    "            warn = True\n",
    "        else:\n",
    "            reference.references.append((cols[0].rstrip(\";\"), cols[1].rstrip(\".\")))\n",
    "    if warn:\n",
    "        import warnings\n",
    "        from Bio import BiopythonParserWarning\n",
    "\n",
    "        warnings.warn(\"Possibly corrupt RX line %r\" % value, BiopythonParserWarning)\n",
    "\n",
    "\n",
    "def _read_cc(record, line):\n",
    "    key, value = line[5:8], line[9:].rstrip()\n",
    "    if key == \"-!-\":  # Make a new comment\n",
    "        record.comments.append(value)\n",
    "    elif key == \"   \":  # add to the previous comment\n",
    "        if not record.comments:\n",
    "            # TCMO_STRGA in Release 37 has comment with no topic\n",
    "            record.comments.append(value)\n",
    "        else:\n",
    "            record.comments[-1] += \" \" + value\n",
    "\n",
    "\n",
    "def _read_dr(record, value):\n",
    "    cols = value.rstrip(\".\").split(\"; \")\n",
    "    record.cross_references.append(tuple(cols))\n",
    "\n",
    "\n",
    "def _read_pe(record, value):\n",
    "    pe = value.split(\":\")\n",
    "    record.protein_existence = int(pe[0])\n",
    "\n",
    "\n",
    "def _read_kw(record, value):\n",
    "    # Old style - semi-colon separated, multi-line. e.g. Q13639.txt\n",
    "    # KW   Alternative splicing; Cell membrane; Complete proteome;\n",
    "    # KW   Disulfide bond; Endosome; G-protein coupled receptor; Glycoprotein;\n",
    "    # KW   Lipoprotein; Membrane; Palmitate; Polymorphism; Receptor; Transducer;\n",
    "    # KW   Transmembrane.\n",
    "    #\n",
    "    # New style as of 2014-10-01 release with evidence codes, e.g. H2CNN8.txt\n",
    "    # KW   Monooxygenase {ECO:0000313|EMBL:AEX14553.1};\n",
    "    # KW   Oxidoreductase {ECO:0000313|EMBL:AEX14553.1}.\n",
    "    # For now to match the XML parser, drop the evidence codes.\n",
    "    for value in value.rstrip(\";.\").split(\"; \"):\n",
    "        if value.endswith(\"}\"):\n",
    "            # Discard the evidence code\n",
    "            value = value.rsplit(\"{\", 1)[0]\n",
    "        record.keywords.append(value.strip())\n",
    "\n",
    "\n",
    "def _read_ft(record, line):\n",
    "    name = line[5:13].rstrip()\n",
    "    if name:\n",
    "        if line[13:21] == \"        \":  # new-style FT line\n",
    "            location = line[21:80].rstrip()\n",
    "            try:\n",
    "                isoform_id, location = location.split(\":\")\n",
    "            except ValueError:\n",
    "                isoform_id = None\n",
    "            try:\n",
    "                from_res, to_res = location.split(\"..\")\n",
    "            except ValueError:\n",
    "                from_res = location\n",
    "                to_res = \"\"\n",
    "            qualifiers = {}\n",
    "        else:  # old-style FT line\n",
    "            from_res = line[14:20].lstrip()\n",
    "            to_res = line[21:27].lstrip()\n",
    "            isoform_id = None\n",
    "            description = line[34:75].rstrip()\n",
    "            qualifiers = {\"description\": description}\n",
    "        if from_res == \"?\":\n",
    "            from_res = UnknownPosition()\n",
    "        elif from_res.startswith(\"?\"):\n",
    "            position = int(from_res[1:]) - 1  # Python zero-based counting\n",
    "            from_res = UncertainPosition(position)\n",
    "        elif from_res.startswith(\"<\"):\n",
    "            position = int(from_res[1:]) - 1  # Python zero-based counting\n",
    "            from_res = BeforePosition(position)\n",
    "        else:\n",
    "            position = int(from_res) - 1  # Python zero-based counting\n",
    "            from_res = ExactPosition(position)\n",
    "        if to_res == \"\":\n",
    "            position = from_res + 1\n",
    "            to_res = ExactPosition(position)\n",
    "        elif to_res == \"?\":\n",
    "            to_res = UnknownPosition()\n",
    "        elif to_res.startswith(\"?\"):\n",
    "            position = int(to_res[1:])\n",
    "            to_res = UncertainPosition(position)\n",
    "        elif to_res.startswith(\">\"):\n",
    "            position = int(to_res[1:])\n",
    "            to_res = AfterPosition(position)\n",
    "        else:\n",
    "            position = int(to_res)\n",
    "            to_res = ExactPosition(position)\n",
    "        location = FeatureLocation(from_res, to_res, ref=isoform_id)\n",
    "        feature = FeatureTable(\n",
    "            location=location, type=name, id=None, qualifiers=qualifiers\n",
    "        )\n",
    "        record.features.append(feature)\n",
    "        return\n",
    "    # this line is a continuation of the previous feature\n",
    "    feature = record.features[-1]\n",
    "    if line[5:34] == \"                             \":  # old-style FT line\n",
    "        description = line[34:75].rstrip()\n",
    "        if description.startswith(\"/FTId=\"):\n",
    "            # store the FTId as the feature ID\n",
    "            feature.id = description[6:].rstrip(\".\")\n",
    "            return\n",
    "        # this line is a continuation of the description of the previous feature\n",
    "        old_description = feature.qualifiers[\"description\"]\n",
    "        if old_description.endswith(\"-\"):\n",
    "            description = \"%s%s\" % (old_description, description)\n",
    "        else:\n",
    "            description = \"%s %s\" % (old_description, description)\n",
    "\n",
    "        if feature.type in (\"VARSPLIC\", \"VAR_SEQ\"):  # special case\n",
    "            # Remove unwanted spaces in sequences.\n",
    "            # During line carryover, the sequences in VARSPLIC/VAR_SEQ can get\n",
    "            # mangled with unwanted spaces like:\n",
    "            # 'DISSTKLQALPSHGLESIQT -> PCRATGWSPFRRSSPC LPTH'\n",
    "            # We want to check for this case and correct it as it happens.\n",
    "            try:\n",
    "                first_seq, second_seq = description.split(\" -> \")\n",
    "            except ValueError:\n",
    "                pass\n",
    "            else:\n",
    "                extra_info = \"\"\n",
    "                # we might have more information at the end of the\n",
    "                # second sequence, which should be in parenthesis\n",
    "                extra_info_pos = second_seq.find(\" (\")\n",
    "                if extra_info_pos != -1:\n",
    "                    extra_info = second_seq[extra_info_pos:]\n",
    "                    second_seq = second_seq[:extra_info_pos]\n",
    "                # now clean spaces out of the first and second string\n",
    "                first_seq = first_seq.replace(\" \", \"\")\n",
    "                second_seq = second_seq.replace(\" \", \"\")\n",
    "                # reassemble the description\n",
    "                description = first_seq + \" -> \" + second_seq + extra_info\n",
    "        feature.qualifiers[\"description\"] = description\n",
    "    else:  # new-style FT line\n",
    "        value = line[21:].rstrip()\n",
    "        if value.startswith(\"/id=\"):\n",
    "            qualifier_type = \"id\"\n",
    "            value = value[4:]\n",
    "            assert value.startswith('\"')\n",
    "            assert value.endswith('\"')\n",
    "            feature.id = value[1:-1]\n",
    "            return\n",
    "        elif value.startswith(\"/evidence=\"):\n",
    "            value = value[10:]\n",
    "            assert value.startswith('\"')\n",
    "            if value.endswith('\"'):\n",
    "                value = value[1:-1]\n",
    "            else:  # continues on the next line\n",
    "                value = value[1:]\n",
    "            assert \"evidence\" not in feature.qualifiers\n",
    "            feature.qualifiers[\"evidence\"] = value\n",
    "            return\n",
    "        elif value.startswith(\"/note=\"):\n",
    "            value = value[6:]\n",
    "            assert value.startswith('\"')\n",
    "            if value.endswith('\"'):\n",
    "                value = value[1:-1]\n",
    "            else:  # continues on the next line\n",
    "                value = value[1:]\n",
    "            assert \"note\" not in feature.qualifiers\n",
    "            feature.qualifiers[\"note\"] = value\n",
    "            return\n",
    "        # this line is a continuation of the description of the previous feature\n",
    "        keys = list(feature.qualifiers.keys())\n",
    "        key = keys[-1]\n",
    "        description = value.rstrip('\"')\n",
    "        old_description = feature.qualifiers[key]\n",
    "        if key == \"evidence\" or old_description.endswith(\"-\"):\n",
    "            description = \"%s%s\" % (old_description, description)\n",
    "        else:\n",
    "            description = \"%s %s\" % (old_description, description)\n",
    "        if feature.type == \"VAR_SEQ\":  # see VARSPLIC above\n",
    "            try:\n",
    "                first_seq, second_seq = description.split(\" -> \")\n",
    "            except ValueError:\n",
    "                pass\n",
    "            else:\n",
    "                extra_info = \"\"\n",
    "                # we might have more information at the end of the\n",
    "                # second sequence, which should be in parenthesis\n",
    "                extra_info_pos = second_seq.find(\" (\")\n",
    "                if extra_info_pos != -1:\n",
    "                    extra_info = second_seq[extra_info_pos:]\n",
    "                    second_seq = second_seq[:extra_info_pos]\n",
    "                # now clean spaces out of the first and second string\n",
    "                first_seq = first_seq.replace(\" \", \"\")\n",
    "                second_seq = second_seq.replace(\" \", \"\")\n",
    "                # reassemble the description\n",
    "                description = first_seq + \" -> \" + second_seq + extra_info\n",
    "        feature.qualifiers[key] = description\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UP_Entry\tID_Enzyme\tName\n",
      "Q9Y5Q5\t3.4.21.-\tInformationNotEspecific... see at https://enzyme.expasy.org/EC/3.4.21.-\n",
      "P62837\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n",
      "P62837\t2.3.2.24\t(E3-independent) E2 ubiquitin-conjugating enzyme.\n",
      "P49427\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n",
      "P49427\t2.3.2.24\t(E3-independent) E2 ubiquitin-conjugating enzyme.\n",
      "P61077\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n",
      "P61077\t2.3.2.24\t(E3-independent) E2 ubiquitin-conjugating enzyme.\n",
      "P51668\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n",
      "P51668\t2.3.2.24\t(E3-independent) E2 ubiquitin-conjugating enzyme.\n",
      "P61086\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n",
      "P61085\t2.3.2.23\tE2 ubiquitin-conjugating enzyme.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import urllib\n",
    "from Bio import ExPASy\n",
    "from Bio import SwissProt\n",
    "from Bio.ExPASy import Enzyme\n",
    "\n",
    "# Las Entry de UniProt a buscar su información como enzima\n",
    "Entrys_UP = ['Q9Y5Q5', 'P62837', 'P49427', 'P61077', 'P51668', 'P61086', 'P61085']\n",
    "\n",
    "# Header\n",
    "print('UP_Entry\\tID_Enzyme\\tName')\n",
    "\n",
    "# Recorre los Entrys para buscarlos en SwissProt\n",
    "for Entry in Entrys_UP:\n",
    "    with ExPASy.get_sprot_raw(Entry) as handle_UP:\n",
    "        record_UP = read(handle_UP)\n",
    "    \n",
    "    # Obtiene su EC number con expresiones regulares\n",
    "    ECs = re.findall('EC=([\\d.-]*)', record_UP.description)\n",
    "    \n",
    "    # Recorrerá cada resultado encontrado en la descripción\n",
    "    for EC in ECs:\n",
    "        # Abrirá su pagina correspondiente en Ezyme\n",
    "        with urllib.request.urlopen('https://enzyme.expasy.org/EC/{}.txt'.format(EC)) as handle_EZ:\n",
    "            with io.TextIOWrapper(handle_EZ, encoding='cp1252') as wrapped_handle:     \n",
    "                record_EZ = Enzyme.read(wrapped_handle)\n",
    "        \n",
    "        # Algunas proteínas pueden no tener una clasificación específica por lo que\n",
    "        # Si existe la página imprimirá los resultados\n",
    "        if record_EZ:\n",
    "            print(f'{Entry}\\t{record_EZ[\"ID\"]}\\t{record_EZ[\"DE\"]}')\n",
    "        # Si no existe, mandará la pagina para buscar más información\n",
    "        else:\n",
    "            print(f'{Entry}\\t{EC}\\tInformationNotEspecific... see at https://enzyme.expasy.org/EC/{EC}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"Ejercicio_3\"></a>\n",
    "\n",
    "## <span style=\"color:yellow\"> Ejercicio_3</span>\n",
    "Escojan un organismo de los presentes en Cellosaurus y después escojan una categoría para graficar los porcentajes.\n",
    "\n",
    "Por ejemplo, en la categoría de enfermedades cuáles son las más representadas o en sexo cuáles son los porcentajes (x% mixed sex, y% female, z% male)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">Para esta actividad se decidió plotear las lineas celulares decidas a la trucha arcoíris *(Oncorhynchus mykiss)* de acuerdo al sexo al que corresponde. Esta especie está bajo una fuerte ameneza patológica y es de importancia económica en el país, aparte, es el organismo que estoy usando para mi proyecto de celular.</span>\n",
    "### <span style=\"color:blue\">*(Oncorhynchus mykiss)*</span>\n",
    "![Aqui está la trucha](https://beneficioss.com/wp-content/uploads/2018/06/1200px-Oncorhynchus_mykiss_opt.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['11B', '23B2', '24BC', '25B', '26B', '49B2', '58B2C', '5C2', '6B [Oncorhynchus]', 'Fin-4', 'Fr994', 'HK1532', 'mRTP1B', 'mRTP1E', 'mRTP1F', 'mRTP1K', 'mRTP2A', 'RBTE 45', 'RT [Oncorhynchus]', 'RT-milt5', 'RT-ovf1', 'RT-testis', 'RTBrain', 'RTE [Carp embryo]', 'RTE [Carp eye]', 'RTE-2', 'RTee', 'RTG-2', 'RTG-2/f', 'RTG-P1', 'RTgill-W1', 'RTgill-W1-pf', 'RTgutF', 'RTgutGC', 'RTH', 'RTH-149', 'RTH1B1A', 'RTH1B1D', 'RTH1B2A', 'RTH1B2C', 'RTHDF', 'RTK [Carp kidney]', 'RTL-4', 'RTL-W1', 'RTM', 'RTO', 'RTO-2', 'RTP-2', 'RTP-91E', 'RTP-91F', 'RTS', 'RTS11', 'RTS11-GFP', 'RTS34', 'RTS34st', 'RTS34st(neo)', 'RTT', 'SOB-15', 'STE-137', 'TPS', 'TPS-2', 'TSS'] \n",
      "\n",
      "Numero de lineas: 62\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import cellosaurus \n",
    "\n",
    "# Parseamos el archivo\n",
    "with open('../tmp/cellosaurus.txt') as handle_cell:\n",
    "    records_cell = cellosaurus.parse(handle_cell)    \n",
    "    # Obtenemos los ID de cada linea celular de la trucha arcoíris\n",
    "    cellsLines = [record['ID'] for record in records_cell if 'Oncorhynchus mykiss' in record[\"OX\"][0]]\n",
    "\n",
    "# Las vemos en pantalla\n",
    "print(cellsLines, '\\n\\nNumero de lineas:', len(cellsLines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 6 42\n"
     ]
    }
   ],
   "source": [
    "# Conteo por sexo\n",
    "Male = 0\n",
    "Female = 0\n",
    "Unspecified = 0\n",
    "\n",
    "with open('../tmp/cellosaurus.txt') as handle_cell:\n",
    "    records_cell = cellosaurus.parse(handle_cell)\n",
    "    for record in records_cell:\n",
    "        if 'Oncorhynchus mykiss' in record[\"OX\"][0]:\n",
    "            if record['SX'] == 'Female':\n",
    "                Female += 1\n",
    "            elif record['SX'] == 'Male':\n",
    "                Male += 1\n",
    "            else:\n",
    "                Unspecified += 1\n",
    "    print(Female, Male, Unspecified)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22.580645161290324 9.67741935483871 67.74193548387096\n"
     ]
    }
   ],
   "source": [
    "# Sacamos porcentajes\n",
    "Male = (Male*100)/len(cellsLines)\n",
    "Female = (Female*100)/len(cellsLines)\n",
    "Unspecified = (Unspecified*100)/len(cellsLines)\n",
    "\n",
    "print(Female, Male, Unspecified)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:green\">El ploteo se generó con el código que proporciona la librería matplotlib.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAV0AAAD3CAYAAAC+eIeLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nO3deXxcdb3/8dfnzExmMkkmbZKm+0YbWmgHWgoERTZBQNACoqIgFgUULqLc61WxerWuP1yuXLd7BVmMooALahEB2Xda2qYlXQgt3dt0yTrZZznf3x/npJ2mSdtsc7J8no/HPDJz5iyf2d75zvec+R4xxqCUUiozLK8LUEqpkURDVymlMkhDVymlMkhDVymlMkhDVymlMkhDVymlMijjoSsiZ4lIZaa3299E5FwR2dnf82aSiCwRkQe8rqOvRGSaiBgR8Wdwm4PyNT0SEdkqIhd0c986ETk3wyWNSAMWut29wMaYl4wxswZquypzROQ6EakQkRYR2SMi/ycio7yuS/WcMWaOMeZ5r+sYCbR7YZDLZOutJ0Tki8APgC8B+cAZwFTgKRHJymAdg/L5Uf1jOL6+XnQvHPK1zG0R/6eIvCkiDSLysIiE0u7/gIisFpF6EXlVRE5Ku+92EXlHRBpFZL2IXJF230wRecFdZ7WIPHyEmt7jrrteRHaIyHXu9KCI/FhEtovIXhH5lYhkd7MOIyIz027/RkS+2828R6r7OhF5RUTuFJFaYIk7/dMiskFE6kTkSRGZ6k4Xd9597mN9U0TmdrPd6e5z0igiTwFFne4/I+15WNPd100RiQDfAm41xjxhjEkYY7YCH8UJ3k+48y0RkT+KyG/dba4TkVPT1jNZRB4Rkf0iUiMiv3CnWyLydRHZ5j6u34pIvntfR1fC9SKyHXg2rbRr3NeqWkS+5s4/zm2JF6Ztd4G7zYD7fL/svs51IrJFRN6fNm+BiNwvIrvd+//W6bn4oltjlYh8Km368yJyQ6fX9eVevGbPi8h33delSUQeFZFCEfm9iMRE5A0RmebO+0sR+e9Oyz8qIrd1sd7Z7mP9mHv7wDdTETldRFa4698rIj9xp4dE5AH3tap3tz22m7q3ishX3fd3nfscpn+ubxSRTSJSKyJLRWRC2n1GRG4RkY3Axi7W3W0dIpIvIve6r8cu97nzuff9n4j8OW09PxCRZ0REjlZTvzLGDMgF2Apc0MX0c4GdneZbDkwACoANwE3ufacA+4BSwAcscucPuvd/xF3OAq4CmoHx7n0PAl9z7wsB7+mmzilAI/BxIAAUAvPc+/4HWOrWlQc8Cvy/bh6HAWam3f4N8N1u5j1S3dcBSeBWwA9kA5cDm4AT3GlfB151578IWAmMAsSdZ3w3j/U14CdAEDjbfdwPuPdNBGqAS9y63ufeHtPFei52a/R3cV8Z8KB7fQnQ5q7TB/w/4HX3Ph+wBrgTyEl/jYBPu4/3OCAXeAT4nXvfNPe5/q27XHbatF+7t08G2oET3GX+CdycVuOdwM/Tnu8EcKNb083AbkDc+x8DHgZG47w/zkl7TZPAt93plwAtwGj3/ueBG9K2eR3wci9es+fd52IGzjeK9cDbwAXue+G3wP3uvKe7tVvu7SK3prHpn0mcz9V24ANdfV5x3ifXutdzgTPc65/F+QyE3edqARA5wud/LTAZ5/PzCgc/D+8Fqt06gsDPgRc7fZaecpfL7mLd3dYB/A24C+e9UYyTLZ917wu7z911wFluDZOOpaZ+zcZBErqfSLv9Q+BX7vX/A77TaflK3Dd+F+teDVzmXv8tcHfHk3qEOr8K/LWL6YIThjPSpr0L2NLN4zjm0D1K3dcB2zvd/zhwfdptC+fDNNV9s7yN8/XeOsI2puCERE7atD9wMHS/ghtsafc/CSzqYl2fAPZ0s507gKfc60uAp9PuOxFoTXsu99N1cD8D/Fva7Vk4wejnYMAel3Z/x7RJadOWAx9zr18FvOJe9wF7gNPTnu9NacuF3XWNA8YDNm6QdvE+bk2vH6eB0BFQz9N96B7Ta5a2nq+l3f5v4PG02x8EVqfd3gC8z73+OeCfnT5r3wJ2Aud193kFXnTnK+o0z6eBV4GTjlRz2vpuSrt9CfCOe/1e4Idp9+W6r++0tM/Se4+w7i7rAMbi/LPNTpv2ceC5tNunA7XANuDjadOPWFN/XgZLn+6etOstOA8YnFD5ovsVol5E6nH+c04AEJFPysGuh3pgLge/Mn8ZJziXi/O19tPdbHsy8E4X08fgfABXpq3/CXd6nxylboAdnRaZCvw0bf5anMc20RjzLPAL4JfAXhG5W5yv/51NAOqMMc1p07Z12sZHOj3X78EJns6qgSLpur9tvHt/h86vbchdbjKwzRiT7KbW9Nq24QRu+lfZzs9RV9vqeB/9HThRRI7DacE3GGOWd7WcMabFvZrr1lhrjKnrYlsANZ3qT99mt3rwmnXYm3a9tYvb6dssw+3ecf/+rtO6bsL5lvTcEbZ3PXA88Jb71f0D7vTf4fwjfsjtbvmhiASOsJ7012gb7ueWTq+vMaYJ51vVxG6W7ay7OqbifOuoSnsP34XT4u3Y1nJgM87n549p6zyWmvrFYAnd7uwAvmeMGZV2CRtjHhSnT/PXOP/NC40xo3C+zgiAMWaPMeZGY8wEnK8j/ytpfa6dtjGji+nVOG/oOWnbzjfGdPehasEJ6Q7juprpaHW7TBc1frbT85BtjHnVfaw/M8YsAObgfFi+1MWmq4DRIpKTNm1Kp238rtM2cowxd3SxrtdwWhQf6vTYcoD347RUj2YHMKWb4N6N8wFKrzPJoWHT+TnqljGmDecDdg1wLYcH0ZFqLJDeHZHRzBHeD8f4mvXGA8BlInIyTrfF3zrdfxPO835ndyswxmw0xnwcJ6x+APxZRHKM03f/LWPMicC7gQ8AnzxCLZPTrk/BeV2h0+vrvm8KgV3pZRyhvu7q2IHzvixKew9HjDFz0rZ1C073wW6chlmHY6mpXwx06AbcTu+OS0/3RP4auElESsWRIyKXikgeTp+NwfmKijg7MQ7sjBCRj4jIJPdmnTtvqott/B64QEQ+KiJ+cXZSzDPG2O727xSRYnedE0Xkom5qXQ1cLSI+EbkYOKeb+Y5Ydzd+BXxVROa4y+SLyEfc66e5z08A54Pe1tXjNMZsA1YA3xKRLBF5D85X0w4PAB8UkYvcxxASZ6fnpC7W1YDz9fPnInKxODukpgF/wvnqeiyhthznH8Ed7usaEpEz3fseBP5dnB1/ucD3gYe7aRUfq9/ifMVf6D7WozLGVOF07fyviIx2H+fZx7i91cCHRCTs/rO/vuOOY33NesMYsxN4A+c1+IsxprXTLI04ffJni0hX/1ARkU+IyBj3M1DvTk6JyHkiEnV3TMVwvn4fqe5bRGSSiBQAi3H6xsHp1vqUiMwTkSDO67vMODtjj6q7OtzX61/Af4tIRJwdsjNE5Bx3ueOB7+J8A7gW+LKIzOuPmnpioEP3nzitxY7Lkp4sbIxZgbOD4xc4wbkJ54ODMWY9Tv/WazgtoChOZ32H04BlItKEszPsC8aYLV1sYztOf9MXcb62r8bZEQNOP+cm4HURiQFP4/QvduULOCFWj9Oi6tzC6Nje0eruapm/4rQ4HnLrWIvTogSI4PxzqMP5elQD/LibVV2Ns1OyFvgmThB1bGMHcBnOh2M/TqvhS3TzHjHG/NCd98c4b/xl7jLnG2Paj/R43OVTOM/XTJydOjtx+l4B7sMJjReBLTihdOvR1nmU7b2C0z+7qocfpGtxPtRv4fTZHnYkQDfuBOI4r3EZzj/3Dj15zXqjDOd91eU/P2NMPU43y/tF5DtdzHIxsM797PwUp2+8Dae1/mec13sD8AJH/gf2B5wQ3Oxevutu/xngv4C/4PzjnQF8rAeP70h1fBLIwtnhWOfON95t8D0A/MAYs8YYsxHn/fs7EQn2Q03HrGMPrVLDnog8C/zBGHOP17UMJLc1/gDOTiDboxq24uxIfNqL7Q9mw+7AY6W6IiKn4RwOdJnXtQwkt8viC8A9XgWuOrLBviNNqT4TkTKcrqHbjDGNXtczUETkBJzurfE4x5irQUi7F5RSKoO0pauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmkoauUUhmko4ypjJp2+2NhnNPudHUpxhnk3e9eGrfecenlHpWq1IDQ0FUDwg3XucBJOIPCn+TeLujBaqqPPotSQ4uGruqzabc/JsB8nLMRnIZzltYZIqLdV0p1oqGremXa7Y9NBC4ELjTGXCAi6WczRkS6XlCpEU5DVx0TtzV7FnCFMeYid8BsQANWqZ7Q0FVHNO32x2YYYz6JMdeJZU0BDVml+kJDVx1m2u2PRYCPGjt1vVi+M0QENGiV6hcauuqAabc/Nt/YqS8jcoWIFRTL53VJSg07GrqKqV9+9GyTSnzXCgTP0qBVamBp6I5gU7+89AMmlfy2FQjOFyvodTlKjQgauiPMtNsfs0wy8TGD+ablzzpeW7ZKZZaGbhoRSQEVaZMuN8ZsHaBtXQecaoz53ECsvytTbvvj2fh8d1mB0GzdLaaUNzR0D9VqjJnndRH9bfKtD0wEudcK51+kh3sp5S0N3aMQER9wB3AuEAR+aYy5S0TOBb4F7AXmAY/gtJK/AGTjtJLfEZEPAl8HsoAa4BpjzN5O2xgD/AqY4k66zRjzSl9rn3b7Y4FkY813feH828Tnz+rr+pRSfaehe6hsEVntXt9ijLkCuB5oMMacJiJB4BUR+Zc7z8nACUAtsBm4xxhzuoh8AbgVuA14GTjDGGNE5Abgy8AXO233p8CdxpiXRWQK8KS73l6bdEvZh61gzs/9eYXj+rKeoSRaFj0O+BeQci/NQF3apdb9WwW8A2yuWFSxt+u1KTUwNHQP1VX3woXASSLyYfd2PlACxIE3jDFVACLyDs4HHpwW73nu9UnAwyIyHqe1u6WL7V4AnJj21T8iInnGmMaePoBx1/54tD9S/JA/r+jCni47VEXmRwqAU3OjuZMKzy+c0ZNlo2XRZpx/mJuBtcBKYEXFoood/V+pUhq6x0KAW40xTx4y0eleaE+bZKfdtjn43P4c+IkxZqm7zJIutmEB7zLGtPal0HFX33Fl1tjjfm0Fc0b3ZT1DUBS4OVGdML1YNsddPgpcdmCFZdF9OAH8BvAS8HLFooq2fqhVjXAaukf3JHCziDxrjEmIyPHArh4sn582/6Ju5vkX8DngRwAiMs8Ys7qbeQ8z6j3XZIVnvfu+4KQ5V4tljdQ9ZQ0SkKZ+XF8x8H73AtAaLYu+hPNaPVWxqOLNftyWGkE0dI/uHmAasEqc7//7gZ6czWAJ8CcR2QW8DkzvYp7PA78UkTdxXpMXgZuOZeXFH1kyP3fexY/4cwum9aAm1XPZuENZAkTLolXA34AHcVrBvWllqxFIjNH3ylAULimVyBkf+Wpw7Ixvij9ruB6ZUL31jkvHHG2myPzIOcCnQlNCTWM/NPaWDNTV2U7gYeDBikUVKz3YvhpCtKU7BOXMOTc3/8yrHw6Om3mJ17UowNlZ+kXgi9GyaCVwF3B/xaKKem/LUoORnk5liBn1nquPH3Xm1as0cAetWcBPgJ3Rsuhd0bJo1OuC1OCioTuEFJx/4/l58y95JVAwscTrWtRR5QCfAd6MlkVfjJZFF3pdkBocNHSHgHBJqRRcePNnck963999OaOLjr6EGmTOAv4eLYuu1PBVGrqDXLik1B8uede3c0+68OdWMCfH63pUn5yCE74romXRD3pdjPKGhu4gFi4pzc4uOeMnOXPO+6o1fI9QGIkWAEujZdHl0bLoe7wuRmWWhu4gFS4pzQ1Nm//z3DnvvVl8fh30dng6DXgpWhZ9KFoWnXLUudWwoKE7CIVLSnNCU076Sd68939SfH49rG/4uwrYEC2LfjVaFtVvNMOchu4gEy4pDQcnzflh3vxLF4nPH/C6HpUxYeD7OEc7nOl1MWrgaOgOIuGS0uzgxBO+H1nwwRvEH9AWz8g0C3gxWhb9YbQsqieuG4Y0dAeJcElpMGv88d/OW7Dws8P4Z73q2FjAl4AV0bLofK+LUf1LQ3cQCJeUZvkixV+OLFh4sxUIhryuRw0ac4Fl0bLof0XLorozdZjQ0PVYuKTUhy9wU37plbdYwbAeh6s6CwDfBp6MlkX1hzHDgIau9xbml155iz8yZqzXhahB7XxgVbQseprXhai+0dD1ULik9NScOed9JTj++OO9rkUNCZNxjuu90etCVO9p6HokXFI6JThpzpLw8Wdqy0X1RBC4O1oW/XW0LKrHcA9BGroeCJeU5vtHjf963ikfeK9Ylr4GqjduAB6NlkV1P8AQox/4DAuXlAawfDdHSq+8zAoEs72uRw1pFwPPRcuiRz27hho8NHQzKFxSKsBH806++KP+3IJir+tRw8JpwKvRsuhxXheijo2GbmZFA2OmXxWaNv8krwtRw8pMnOA92etC1NFp6GZIuKQ0D8v/mcipC88Qy9ID3VV/Gws8HS2LzvW6EHVkGroZ4HYrfDzvlEvP9IXztf9NDZQi4JloWfQErwtR3dPQzYx5WWNnXhaaEtVuBTXQinGCV8+jN0hp6A6wcEnpKPEFPpO34IOlInp4mMqI8cCzunNtcNIQGEBut8I1OdH3neLLziv0uh41okzCGa9B33eDjIbuwDrFyo6cnT3tZN2rrLwwE/irno1icNGfEQ6QcElpEPhE3imXlogvkNHBqJOx/VQ/9hNSTXWIWOTOu4jIqZdR99x9tGxajvj8+EeNo+iS27BCuYctb7c1UfP4z4hXbweg6JIvEJx4AnXP30/r5pVkFU+n6ANfBKBp7bPYbY1ETr0skw9RHbuzgHuAT3pdiHJoS3fgnOMvmDgla+yMzLdyLR+jz7ueiTf+inHX/pjGVY8Rr95OaNo8Jlz/SyZ8+hcECibS8Pqfuly89pm7CR23gIk3/ooJn/45gcLJ2O3NtO/awIRP/wJjbOL7t2In2mle+zR58y/N8ANUPXRttCz6Da+LUA4N3QEQLimNAFfmzbtkrhc7z/y5BQTHzQTACoYJFE4m1VhD9vRTEPcQ4eCEWSQbqw9b1m5voW3HOnJPuhAA8QXc1rBgUkmMMZhkHLF8xJY/Qt6ChYhPvzANAd+KlkWv8roIpaE7UC4JTjpxYmD0eM+HbEw27CW+dzPBCbMOmd705lNkH3fq4fPX78EXjlDzz/9h9/2fp+bxn2HH27CCYcKz3k3Vbz6PP38sEswhXvU24ZIzMvVQVN/dEy2Lev6eHOk0dPtZuKR0PHBh7twLPD+3lR1vZf9fv0/B+TdiBcMHpje8+jBYPnJOPPewZYydIr7nHfLmX8KET/0MCQSJud0Q+aUfZsKnfk7Be2+g4aUHGHXWJ2hc8yT7/3YH9a8+lKmHpXovF/hjtCyqp4TykIZuP3IPEftwaPop43w5oyZ4WYtJJdn/1++Tc+K5hGe9+8D0popnaHlnOUUf/E9E5LDl/HlF+PKKDrSMw7POJL73nUPm6bjtHz2R5rXPMuby20ns30aidtcAPiLVT04G7vS6iJFMQ7d/zQAWhGeeMeuocw4gYww1j/+UQOFkIqdfcWB66+aVxJb9meIrv4EV6Lqx48sdjT9SRKJmJwBt29YQKJpyyDz1Lz1A/nuuATsJxnYmioVJtg/MA1L97aZoWfQjXhcxUukekP51SaB4erY/UuTpL4Had62ned1zBMZMY/f9twIw+uxPUvv03ZhUgr0Pfx1wdqYVXvQ5ko011DzxM8Z+5FsAFFxwE9X/+DEmlcQ/ahyFl9x2YN0tb79G1rgS/O5vPYITZrP73lsIFE8jq1h/ADWE3BMtiy6vWFSxzetCRhoxxnhdw7AQLikdB9wx6pxFJ2cVTZ3ndT3DRPXWOy496gBBkfmRc4BPhaaEmsZ+aOwtGahruHiyYlHFxV4XMdJo90L/Oc+XM9ofKJgc9boQpY7RRdGy6LVeFzHSaOj2g3BJaR5wfs7c86foWLlqiLkzWhbVs5hkkIZu/3i3+LMCwXEzF3hdiFI9VAj8zOsiRhIN3T4Kl5RmAZeGZ72nSPxZ4aMuoNTgc1W0LPoBr4sYKTR0++5kIBKcMMvTw8SU6qM7dTSyzNDQ7bv3WeFR7b68wuleF6JUH8wEbvW6iJFAQ7cPwiWlBUBJuOSMiXpWCDUM/Fe0LFrgdRHDnQZF35wEmKxxM/UMrGo4yAe+5nURw52Gbt+c68srSvpyRk85+qxKDQm3RMuiU70uYjjT0O2lcElpMTA1XHLGZOlq5BilhqYg8FWvixjONHR7LwqYrOLp2rWghpvromXRcV4XMVxp6PaCO4TjeVZ4VKsVHjXJ63qU6mdB4N+9LmK40tDtnbHAxNDkuYXas6CGqZuiZdF8r4sYjjR0e2cGYAJFU3SHgxquIoCO2DYANHR75ySg1Z9fPM3rQpQaQF/QX6n1Pw3dHgqXlFrAXCs70maF8sZ6XY9SA6gYuNzrIoYbDd2eGwtkhybPGa+HiqkR4AavCxhuNHR7bhpAoGiq9ueqkeCCaFl0mtdFDCcauj0XBdr8+WOneV2IUhkgwPVeFzGcaOj2gHt8bhTL16D9uWoEuS5aFtUzovQTDd2eGQPkBIqm5ulpedQIMgk4z+sihgsN3Z4ZB5hAwcSjnqFWqWHmCq8LGC40dHtmHIA/t7DQ60KUyrDLo2VRPVqnH/i9LmCIOQ5oi63428vNb724KlA4udAfKS7y5RUWWdn5hb5QbpEEs0frgOZqGJoAnA4s87qQoU5Dt2cKgTwglGqqbU011dYAOwFzYA7LZwUKJ48OjJ5Q5IuMKfSFRxX5svMKrVCunrhSDXUfQkO3zzR0e+YuYApON8NUYDLOTgZwgteHnYon9m9tSezfugWoTF/YCo/KziqaXOjPH1vkyy0stML5Rb5QbpFkhUfrjjk1BFwBfMXrIoY6Dd0eaNm4rBqoTp8WLin1AaOBIvcyCSeQJwK5pLWC7Zb6trbt9XVQsRuwD6xELAkUTh4dKJhQ5MsbU+jLGVVkZUcKrWBOkRUI5gz4A1Pq2JREy6LTKhZVbPW6kKFMQ7ePWjYuS+EEcXXn+8IlpWGcIC7E+flwR+t4As5B5wawMHYyUb2tJVG9bSvwdvo6rOxIKFA0pdCfP7bI77aOrey8IisrXKCtY+WBc4CtXhcxlGnoDqCWjctagO3u5QC3dTyKg4E8mYOt43EcbB2L3Rprbd+xtr59x9oqDmkdiwQKJo3yj57g7MzraB2HcousQDB3wB+cGqnOAcq8LmIo09D1gNs6rnEvhwiXlGZzMIyLccJ4CjAe5xA/t3VskomaHa2Jmh07gE3p67BCecGO1rEvr7DQF84vskJ5RVYwXCCWT19z1Rdne13AUKcfwEGmZeOyVmCHeznAHVIyvXWc3nc8DkjhdFmI3dbY1r5zXax957o9dGod+0dPzA8UdLSOR7ut45wiKxDKy8TjU0PejGhZdGLFoopdXhcyVGnoDhEtG5fZQK17OUS4pDSEE8RFOK3jKTiBPA7o6PcVjEkla3e2JGt37gTeSV+HFcrNChROLvSPGlfky3Vbx9l5RVYwp1Bbx6qTc4A/eF3EUKUfpmGgZeOyNmCXeznAbR3nc7B1PBEnjCdxaOvYstuaWtt3bWhq37Vhnzv9AP/oCfmBgomF/siYIl9OQZEVjhRawdwiCQQjOqTwiHQKGrq9pqE7jLmt4zr3sjH9vnBJaZDDW8cdfccd7wsBUsm63a3Jut27gc3p65CscCCraIrTOs4rLPSFRxVZobxCKxQuEssfGNAHp7wU9bqAoUxDd4Rq2bisHdjtXg5wh6/srnVcjNNHLICYeEt7++63mtt3v7Wfzq3jUeMigYJJhb7ImCJfbkGRzz2yQgKhfG0dD3kneV3AUKahqw7RsnGZAerdyyFHRYRLSrM42Doew8G+4wk47yXnyAqwk/V7WpL1e6pwjuk88AMRCYT8gaKphYH01nF2XqEVzCkSn19Pgjg0jIuWRYsqFlUcdmy6OjoNXXXMWjYuiwNV7uUAt3Uc4WDreAJOIE92Lx2tY8sk2triVZWt8arKSiCZvh5//tg8f8GkQl/OqBlZxdPrA6MnNA34g1K9FQWe87qIoUhDV/WZ2zpucC+HHBXhto4LOLR1PAUnmIM4gWwBJtmwtyXZsHcvkGiGx1s2Lnskc49C9ZCGbi9p6PbVkvz/wTl+9i2cAW4qgU0saYh7Wtcg4baO97iXA9zWcR4HW8fjOfgz6QDQmNlKVQ9N97qAoUpDtxcWzgoIMAew/3pV9hU+S6Z0miXFkvyOUcY6Lk4oL2nYm9lqBye3dRxzL4ccFREuKQ3QqetBDTqTjj6L6oqGbu/kAP+R5QMRJndxvw+Y6V4uPeSeJfkNdA5i57KRJQ3tA1n0UNGycVnC6xrUUWno9pKGbu8UAPbxhVaL1fPjn/JxRuA/vdN0myX5Wzm0dewE85KGKpQaXCZ6XcBQpaHbOwWAjMuV/hzNy8I5HdBxwPsPuWdJfgxnyMeuWset/ViDUsdqfLQsalUsqrCPPqtKp6HbO6MAa1RIsjO0vQhwqntJZ7Mkfztdt451QBI1kPw4PyXffbQZ1aE0dHunAEiOConX5zyzgGnu5aJD7lmS38TB1nF6C/ltljS0ZLJINWyNQkO3xzR0e2cUkMjNylhLtzdycQYmOaXTdMOS/J0c2k3RcdnBkgaDUsdGTyXVCxq6vRMBEnlZnrd0e0M4+Eux93W6r4Ul+d21jvXXYaozDd1e0NDtnXwgEQ4wFEP3SMLAPPdyqCX5uzh8R14lsE1bxyPWcHv/Z4SGbu9EgHh2QEJeF5JBE93LeztNb2VJ/ka6OvZ4SYP+qmx405ZuL2jo9k4esNcSLK8LGQSycYb6O3y4vyX5VXTdOt7KkgY91Gjo09DtBQ3dHlo4K+ADsug0fqzq0nj3cm6n6W0syd9E1z+TbujtxgKjA5HeLqt6RRsdvaCh23NC2viwqldCwFz3cqgl+XvpunW8hSUNXf6ji5wWmZE9Lfvc4ITg1AGrWHVFxxa6upcAAB0OSURBVMfoBQ1dNdiMdS8dp/o2wFJgCbA6fca5989dYMftH45696jzRE9H4QUN3V7Q0FWDkjGGTbX2W4++nfzT81tT31tamTgwGNDce+e+yyTND6yQdZYv5DvSatTA0gGaekFDVw0qtjFmY429vmxNonLtPrsF2ILTpcPc++aeZZLmB1bQepf4tWE7COi4H72godsHxmjfbn+xjTGV1fba+1cn3nqr2m4EngeeXFqZ2Dfn3jnnz7l7zh1W0DpVfBq2g4iGbi9o6PZBa9K0ZWI7P329nV+vSmCAG08JcNsZwUPu/9Er7fy+whmCNmnDhmqb/V/KI2Ubrni4lfo2w3ffG+Ty2c5Z0S97qIX/uzTEhDzvdz6nbGNvqLYr7itPVG6qtRuBZ4CnllYmqufeO/fiOXfP+b4VtObrO3VQ6vWRJiOZvpV77kDrtiXBgA8cs3Zfil+vSrD8xhyyfHDxAy1cWuKnpPBgX+aXzgzypTOdIH60MsGdr8cpyBZ+tizOopMDfGxugIt/38LlswM8WpnglHE+zwM3ZZvU2n32m/eWx9/eWm+agKdwwrZ2zj1zPuiG7VztRhjU9hx9FtWZhm4PLa1MpBbOCrQB/qa4GfDQ3bDf5oxJPsIBJ3zOmernr28l+fKZXe9AenBtgo/PdVq0AUtoTRraUwZLIGkb/mdZnEc/7t2vN5O2Sb65115z76r4xh0x0wQ8ATyzZfHsBjthf2jOr+d8z8qyZnlWoDpWNrDP6yKGIg3d3mkEAo3tZsD7tOYWW3zt2RQ1LTbZAeGfm5KcOr7rVmpLwvDEpiS/uMQZ/OzqaICrH2nlt2sS/OCCEP/7RpxPnhQ4EOCZlLRNorzKLr+3PL55d6OJAY8Dz21ZPLvRjttXEbe/Y2VZMzJemOqtmopFFXrIWC9o6PZODBgdax/4lu4JY3x85cws3ve7FnKzhJPHWvitrkPz0cokZ07xU5Dt3J8fEh672mnV1rUafvBKO49cFebGpa3UtRm++K4s3jV5YN8CiZSJr6xKld+7KrF5b7OJAf8AXtiyeHazHbc/Qdz+lpVlTRvQItRA0BOs9pKGbu80AMV1bQMfugDXn5LF9adkAbD4mTYmRbpu6T607mDXQmfffqGdr50V5MGKBAsm+Lg6GuCyh1p4btHAvAXiKdP+xq7UqnvLE1uqW0wMeBR4ccvi2W123L6OhP1NK8vSkxsOXdqf20saur0TAwL7m01zJja2r9mmOMdie4PNIxuSvHb94eOMNLQZXtia5IErDh9XfWNNit1NNudM87N6TzvZAUGAtgH4ctieNG2v70ytvH91Ylttq2kA/ga8vGXx7IQdt28gYf+XlWWN7/8tqwzb7nUBQ9WwCF0RMcADxphr3dt+oApYZoz5wBGWOxf4zyPN0406IFBZY9cYYxjoX6Be+cdWaloMAR/88pIQo7OFX62IA3DTqU4L+K9vJbhwhp+crMNr+dqz7Xzvvc7RDR+PBrj8oVZ+uizOt88NHjZvb7UlTeurO1Ir7i+Pb29opw4nbF/dsnh2ym63byZhL7ayrOJ+26Dy2gavCxiqhkXoAs3AXBHJNsa04pwRYSBPzFgH+FoSJFsSxHKyGNDRrV761OEt246w7XDdvCyum5d12HwAf/zIwaMVinMsXu2ipdxbrQnT/NL21Iqy1fEdjXFqccL2tS2LZ2O327eQsG+3glZRv21QDRYaur00XEIXnL3hlwJ/Bj4OPAicBSAipwP/gzP2ayvwKWNMZfrCIpID/ByI4jwvS4wxf+9mW3U4h8zQ0G5qc7JkxA0p2JIwTS9sTa4oW5PY0ZKgBngEWL5l8Wyf3Wb/uyTNl6ygNdrrOtWA0dDtpeEUug8B3xCRf+AMqH0fbujiDBN4tjEmKSIXAN8Hruy0/NeAZ40xnxaRUcByEXnamC77bWtxfyRR3WKqJ+Qxrf8fzuDUHDexZ7ckV/zuzcTutiT7gL8AK7Ysnp1lt9m3S9L8hxWyRtw/oZHEGNMmIlu9rmOoGjaha4x5U0Sm4bRy/9np7nygTERKcMKyq138FwILReQ/3dshYApd/0evxR3AeVfM3n/S2OE/0lVT3DQ89U7yjd9XJHbHU+zD+Uaxasvi2dmpttQ3rKT5vBWy8ryuUw08EamsWFShZ/7opWETuq6lwI9xzlRQmDb9O8Bzxpgr3GB+votlBbiyc7dDlxupTLQtnBVoAIJb6s3+PtY8qMXaTd2Tm5IrHlqbqErYVOGE7eoti2fnpdpS37NS5t98IZ+etmVkWX30WVR3hlvo3gc0GGMq3CMTOuRzcMfadd0s+yRwq4jcaowxIjLfGFN+hG1tA45bsyc1LA8Sr28zNY9vTK7447rE3pRhF07YrtmyePaoVFvqx1bKfMYX8h1+fJoaCV71uoChbFiFrjFmJ/DTLu76IU73wn8Az3az+Hdwdra96Z6FYCtwpEPJNgPRqiZT29BmavJDUniEeYeMulaz/x9vJ1b9ZUNyj23YjtNnW7Fl8eyiVFvqZ5ZtrveFfP13rJkail7xuoChbFiErjEmt4tpz+N2IxhjXgOOT7v7v7qYpxX4bA82uxN3cO2dMXtHfsg3pEO3psXeu7QyuepvbyX3GWfg8EeAdVsWzx6bak39yrLNIl/I1/UxaWrEMMbUi8h6r+sYyoZF6HrkwC9yNtXa2+cU++Z5WUxv7W+2q/72VrL8H28n9xvYhBO2G7Ysnj0p1Zq617LNNb5sX9e/LVYjjoi8VrGoQgfv7wMN3d6rAZqA4Irdqe2XzR5aubS3yd75yIbk6sc3JWtwzrb7V6Byy+LZU1Ntqd9ZtrnKl+3T94fqTLsW+kg/VL20tDJhFs4KrAeia/ba+9uSpiXkF+8Gqj1GVY329j+tT6x5enOqFliH8wuyTVsWzz4u1Zp62DLmSl/I5/0pJdRg9YzXBQx1Grp9UwGcDlDVaHZMHy2DdvDtXTF768PrEhXPb03VAm8Cfwc2b/7KrOPthP2IZcxCX7aGreqeMWa/iCz3uo6hTkO3b7bj/jKtsia1efrowXfGgx0N9jsPrk2sfXl7qh4oxzmWeevmr8w60U7Yj1oWl/hCerbH/pJqTrHr/l207WxDRJh4/USq/1VNvMoZoCjVksIX9jHzOzMPWa69qp0d/7vjwO34/jjFVxRTdFERe/64h8Y3G8meks2kzzijYda9UkeqOUXRhRkd1uIx/VFE32no9k0VkAR8T72Tqrx4ZuD9XhfUYWu9vfH3bybWLduVqgdWAP9YWpnYNvf+uSfbcftxy+JCDdv+V/WHKnKjuUz53BTspI1pN0z5tykH73+wCl/48F8wBscHDwSxsQ2Vt1USWRAh1ZKiZVMLJd8tYcevdtC2o42ssVnUv1zPtC9Oy9TDAkBEHs3oBocpDd0+WFqZSC6cFVgHHL+x1q6ubrGrisLejRVrjGFLvan83Zr4upVVdgx4HXhsaWVi59z755564l0n3mcFrfM0bAdGqjVFc2UzE2+YCIDltw75hBljaHijgelfnn7E9TStbyKrOIusoixSrSlM0mCMwSQM4hOqH6+m8H2FZPKkncaYuIj8K2MbHMY0dPvudWAewPr99ltnT8186NrGmHdq7bd+uyaxYc1eOwa8DDy+tDKxe+69c9914l0n/sEX8p3lCw3/MSK8FN8Xx5/nZ9c9u2jb0Ub2tGzGXzMeK+h0lbe83YI/4ic47si/LWlY1kD+GfkA+LJ9RE6N8M433iHnxByssEXr5laKL8v40MTPVyyqaMr0RocjDd2+e8v9K89vTb519lT/eZnasG2M2Vhjr//N6sSGdfvtRuAlnLDdM/e+uWfPuXvOX6ygdYbPr2GbETa0bmtl/CfGE54Rpur3Vez/x37GXjkWgIbXGxhVOurIq0jaNJY3Mu7D4w5MG3PJGMZcMgaAXfftovhDxdS+UEvT2iZCk0MULxz4ABaRBwd8IyOE7q3uo6WViRiwERi1Yre9L9Zuagd6m7Yx9vr9qTe/8lT7n770VPuydfvtpcDtSysT97/z5ePnzrl7zgrxyQtW0DpjoGtRB/lH+wmMDhCe4Rw5GDk1Qus254TRJmVoWNlAfmn+EdfR9GYToakh/PmHt4c61hUcF6T+lXqm3DKF9p3ttO9p7+dHcihjm1acsTdUP9CWbv94BWcgnbq3qu23Tp/oe/dAbCRlG3v9fvvN+8oTb79TZzfiHDP51NLKRPXce+e+f87dc75vBa15+qp6IzAqQKAwQHtVO8HxQZrWNxGaEAKgaV0TwfFBAgVH/hFNw+sNjDqj69bwvkf2MeG6CZikcYfQByyw4wN8QIHhL9q10H/049k/NuCOw/D4xsTq/g7dlG1SFfvsNfeVxzdurTdNwFM4YVs75945C92wnZPJHSuqa+OvGc+Ou3ZgkoasMVlMusE5xKth2eFdC4m6BLvu38W0/5gGgN1u07SuiQnXTThsvbGVMbKnZxMY7YR29sxsNn59I6FJIbKnDOxgb+KT+wZ0AyOMGKM/o+6rhbMCAnwbyAFid38w9Olxudbkvq43aZvkmj326vvK45t2xEwjzvCTz2xZPLvBTthXYviulTX4jg1Ww4dJmV3ik8k63kL/0ZZuP3B/Evw0ThdD7NUdqZUfOqH3oZu0TWJVlV1+76r45qomE8M5/9tzWxbPbrQT9lXE7e9YWdaMfipfqe5Z3KuB2780dPvPSuBawPfn9Yl1l5b4Lw76JdSTFSRSJr6yKrXqnlWJLfuaTQz4B/DClsWzW+y4/Qni9resLGvqQBSvVGfGmISI/K/XdQw3Grr9ZGllomnhrMArwBlNcarW77ffnD/ed/qxLBtPmfblu1Ir7ytPbK1uMTHgUZywbbfj9nUk7G9aWdakAX0ASnViEubhtTeuHZZnRvGShm7/ehE4G+DRtxMrjxa67UnT9vrO1Mr7yuPb6tpowBle8ZUti2cn7Hb7RhL2160s737hpkYuY4yxsqzve13HcKSh2782A3uAvBW77X27G+1tE/IO7w5oS5rWV7anVvxmdXx7Qzt1OMMrvrpl8WzbbrdvImEvtoJWxn9ypFQHkzDPr71xbVdnwlZ9pKHbj9wdak/g7FBrfOzt5Es3Lsg6ELqtCdP80vbUivvL4zuaE9TitGxf37J4Nna7/TmS9u1W0BrSp/1Rw4P45Tte1zBcaej2v5XA1UDWo28n37l8tr8qJ0vynt+afKNsdWJXa5JqnFPiLF9amWiPzI9EQn/d+6HihcV3iE/097rKc3bcXr3uxnXPeV3HcKXH6Q6AhbMCV+CcSXjHrEJr0rYGO9iWZB/OmXVXLK1MxAEi8yPzgZsBX/HlxSdnT8te4F3VSjnshH3OuhvWveh1HcOVtnQHxnPApcC0yhq7I2xXLa1MJDrNJ0AQ2FL3cl19aHLoZPGJvibKM6mW1Mvrb16vgTuA9AM+AJZWJuoXzgr8GkgBq5dWJpLdzLoG2AXkJ6oTDS2bW17NKck5O2OFKpXGGGPEL7d4Xcdwp90LHovMj8wD/h3YYoWswMRPTfycFbQiXtelRp5Uc+rR9f+2fqHXdQx3OrSj994E3gbG2G12onFN49NeF6RGHpMyCSto3ep1HSOBhq7HYuUxG/g9zmA5Vv2r9RXJhuROj8tSI0yqKfXLtdev3eZ1HSOBhu4gECuPbQWeByYA1L1S97h2+6hMSbWl9iHc7nUdI4WG7uDxN5yhqYMtb7fsbtvR9obXBamRIRVL3bzh1g0De/oJdYCG7iARK4/V45wSZTxA9ePVT6VaU3XeVqWGu0Rd4qnKL1U+crT5RGSaiKztNG2JiPznwFXXOyLyatr1H4nIOhH5UQ+W3yoiRQNTnYbuYPM8sAMotFvtRP0r9X/XbgY1UOx2uznVkrrW6zr6mzEm/cwtnwVOMcZ8yat6OtPQHURi5bE4cA+QC/ib1jZta9vettzjstQwZIyhfU/7rRsXb+zz0I0i8ryI/EBElovI2yJyljvd57Y03xCRN0Xks+708SLyooisFpG1afM3ich/i8gqEXlGRMa402eIyBMislJEXhKR2e70sSLyVxFZ417e3bEe9+9SnB3Uy0TkKhEZIyJ/cet5Q0TOdOcrFJF/iUi5iNyFe+qtgaKhO8jEymPbcAbCmQRQ/UT109rNoPpbfE/88U3f2HR/P67Sb4w5HbgN+KY77XqgwRhzGnAacKOITMcZm+RJY8w84GRgtTt/DrDKGHMK8ELaeu4GbjXGLAD+E+gYWP1nwAvGmJOBU4B16QUZYxYCrcaYecaYh4GfAne69VyJ08DB3c7Lxpj5wFJgSr88I93QX6QNTk/gvEmL7Fa7uvb52r8UXVT0KbF0QBzVd8nG5L6m9U0f6+Fi3fVzdUzv6BdeCUxzr18InCQiH3Zv5wMlwBvAfSISAP5mjOkIXRt42L3+APCIiOQC7wb+JHKgARp0/74X+CSAMSYFNBzlMVwAnJi2noiI5OGMgf0hdz2PiciANnK0pTsIpXUz5ACBlsqWXU1rm/7lcVlqGDBJk2zd1vqx3b/dHevhojXA6E7TCoBq93rH0Q8pDjbmBKeFOs+9TDfG/MsY0zHY/y7gdyLyye7Kxcmo+rR1zDPGnNDD2jtYwLvS1jPRGNOYtq2M0NAdpNxuhoeAyYDUPlu7vH1P+7qjLKbUEbXuaP3p1h9t7fGwjcaYJqBKRM4HEJEC4GLg5SMs9iRws9uiRUSOF5EcEZkK7DPG/Bq4F6drAJw86mgVX43zlT8GbBGRj7jrEBE52Z3nGZxR+jr6j4/28/l/AZ/ruCEi89yrLwLXuNPez+H/XPqVhu7g9hSwHJgIsG/pvqWp5lT1kRdRqmttu9pe3v/3/V/uwyo+CXxdRFYDzwLfMsa8c4T57wHWA6vcw83uwmkFnwusFpFynL7Vn7rzNwNzRGQlTtfBt93p1wDXi8ganH7by9zpXwDOE5EKnG6NOUep//PAqe5OvfXATe70bwFni8gqnC6R7UdZT5/ogDeDXGR+JBf4BhACakNTQ2OKFxbfID7J8rg0NYTEa+I7a56smVf9ZHWN17V0R0SajDG5Xtcx0LSlO8jFymNNwC9x+ndDbdva9te/Uv9no/8t1TFKNadiseWxywZz4I4kGrpDgNu/ez/O2Ay+2KrYxsbVjY95XJYaAuyEnYitit1U9VDVKq9rOZqR0MoFDd2h5BXgMZxjCKXuhbqVLZtaXvK4JjWIGdvYjasbfxBbGXvI61rUQRq6Q0SsPGZwxmZ4HeeIBvb/Y/+z7bvb3/S0MDUoGduY2IrYb+pfqV/ivnfUIKGhO4TEymMp4D5gI+4wkHsf2fv3eE18k6eFqUHFGENsZWxp/av1n3ffM2oQ0dAdYmLlsXbgFzgHqxebpLH3PLTnIQ1e1aFxTeNT9a/UXxcrjzV7XYs6nIbuEBQrjzUCdwJxoMgkTGrPQ3seilfHN3pcmvJY09qml+qer/u4O1SoGoT0ON0hLDI/Mg64HeeA82oJiG/cVeM+mlWUdbzHpakMM8bQuLrxhboX6j4WK4/t8boe1T0N3SHODd6vAAE6gvej4z6SNSZrlselqQwxtrEbXm94tmF5w6di5TE9v94gp6E7DETmR8bitHizgP1YSPHlxZdkT8k+1ePS1AAzKZOqe7HuscY1jbdo4A4NGrrDhBu8XwGygb0AhRcWvjvnhJz3pQ1lp4YRO2HHa5+t/UvzhubbYuWxfV7Xo46Nhu4wEpkfKcQZRHoCzml/yC/NPzH/9PwrxCc6dvIwkmxKNtQ8WfOHth1tX4uVx3SQ+yFEQ3eYicyP5OCcF+pkYBtg55yQM7ngvIKPWVlW2NvqVH9o39u+c/+j++9LNaV+5I7NoYYQDd1hKDI/EsAZj/R8nGHqkoGiQN6YD4z5cGBUYEBPRaIGVtOGprU1T9X8HJvfxspjbV7Xo3pOQ3eYisyPCHAJ8FFgP9CED2vMpWPOz56e/W7t5x1aTNIk6l+rfy22MvZ94KlYecz2uibVOxq6w1xkfuQk3NH1cXewRU6NzMovzb/cClgh7ypTxypRl9hb/UT1s/G98e/FymN69pAhTkN3BIjMj4zBCd7jcHawpbLGZY0qurjoCu1uGLyMMaZ5ffOammdrniDFT/VHD8ODhu4IEZkfycI5Ncr7gT1AC4IUXlB4Rs7snPfq0Q2DS6olVV/zdM2rrZtbHwL+HCuPtXpdk+ofGrojiNvPewpwA+ADdgOEJoUKCy4o+GBgVGCql/Upp3Xbuql1bc0zNa/ZbfYvY+UxHbpzmNHQHYEi8yMFOCcZnI/T6m0FGH3u6AW5c3MvsPza1+uFRH1iV+0ztavadrQ9A5TpoDXDk4buCBWZH7GAM4BrcVq9VYDx5/uzC84vOC80KbRALNFR6DLATtitsVWxFQ2vNawDHgRe1qMThi8N3REuMj8yGvgYTgDXA3UAoSmhotHnjL4oqzBrppf1DWfGNnbrtta1tc/Ubkg1pZ4C/qKt2+FPQ1d19PWeAFwDTMQ5tKwVIG9+3sz8U/Pf58vxFXtY4rBijCG+J76+9rnat+L74m8Dv4mVxyq9rktlhoauOiAyP+IH3oXT8s3G2dGWBIicFpmdd1Le2f48/3gPSxzSjDHE98XfqnupbkP7zvb9wCPAs7HyWMLr2lTmaOiqw7jjN1yM84s2g7OzLQlOyzfv5Lyz9PjeY2dsk2qval/XsKzhnbbtbXXAE8CTsfJYzOvaVOZp6KpuReZHioAL3IvghG8CIHdO7tTck3NLs4qyZukOt67ZcbuldWvryvpX67cn65Nx4Dng8Vh5rNrr2pR3NHTVUbk7294LXIRzaqC9QDtAoDCQm39a/vzQtNACX8iX72GZg0ayMbm7ubL5zYZlDftMwsSB53HGS9Axb5WGrjp2kfmRPOAcnK6HXKARqAUMguTNy5uRe0LugkBhYOZI+4Wb3WY3tO1sezNWHtvSvqvdADHgMeA17UZQ6TR0VY+5PymeA1wIzAZSwD6csxNjha2svJPzjs+enn1iVmHWTPFJwLtqB44dt1vi++KVTeubNjSvb27FObv2DuBRYI3uIFNd0dBVfeKeGPNMnLF7gzgBXE1HAIesQN5JeTND00KzsgqzpltBK+JdtX1jjCHVmNrdvqd9Y8umls0tb7ckcM5L14jThfAGsDNWHtMPleqWhq7qF+7A6TOBU3EOO+sI4Brc/l+A4KRgQXhmeHpwfHB6YHRgmpVl5XhS8DEwxphUc2pfsi65s31P+/amdU27kvXJIM4v+FLAa+5lY6w8lvK0WDVkaOiqfuce7zsDWIATwDk4Rz+04Pzq7cDX7uCkYEFoYmhc1piscf5R/nH+PP9YL1rDxja23WbXpZpS++M18V3tu9p3tmxq2Wu32dk4/de4tS8DKoDNeuYG1RsaumpAuWM8jAWmA/OAKE4rWHBawE04YXxgrAF/vj87ODE4JjAqkO/L8+X7cnz5vrAv3wpZ+VbQyhWfhHp6mJqdtNtM3DTbCbvFtJtmu91uSsaStYmaRHX7vvaa9qr2elJkAxEggHN8chLYiNNt8DawR7sOVF9p6KqMisyP+IDxwBScQdVLcH56DM6OKIA299KO0zd82JvUyrYCvrAvaIWsLCtoBcQnPmMbmxS2sY1tUu4lYZLJhmSLSZqOUPfj/Nqu42K767dwjkNehxOwO4F92m2g+puGrvKc2x0xBhiHE8jjgWL3ko8Tih1vVEn7K+50SZtu0v52Xs7CCfM9OCfs3I5z/rgaoEa7C1QmaOiqQc0N5AiQB4TcSxCnCyDgXrdxugLstEsKZ9CeZpzui2agOVYeS2b4ISh1CA1dpZTKIP3NvFJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZZCGrlJKZdD/Bwwk3oYZYnq1AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Ploteo\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Pie chart, where the slices will be ordered and plotted counter-clockwise:\n",
    "labels = ['Female', 'Male', 'Unespecifed']\n",
    "sizes = [Female, Male, Unspecified]\n",
    "explode = (0.2, 0.1, 0.0)  # only \"explode\" the 2nd slice (i.e. 'Hogs')\n",
    "\n",
    "fig1, ax1 = plt.subplots()\n",
    "ax1.pie(sizes, explode=explode, labels=labels, autopct='%1.1f%%',\n",
    "        shadow=True, startangle=90)\n",
    "ax1.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.\n",
    "\n",
    "plt.title('Lineas celulares de Oncorhynchus mykiss por sexo')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"DONTCLICKME\"></a>\n",
    "## <span style=\"color:yellow\"> CHOMP!</span>\n",
    "![Te dije que no le dieras clic](https://m.gifanimados.com/Gifs-Animales/Animaciones-Animales-Marinos/Peces/Tiburones/Boca-Tiburon-69162.gif)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
